diff -udpr gtk+-2.6.7_/demos/gtk-demo/geninclude.pl.in gtk+-2.6.7/demos/gtk-demo/geninclude.pl.in
--- gtk+-2.6.7_/demos/gtk-demo/geninclude.pl.in	2003-11-09 00:08:04.000000000 +0200
+++ gtk+-2.6.7/demos/gtk-demo/geninclude.pl.in	2020-12-27 16:22:52.529875169 +0200
@@ -40,7 +40,7 @@ foreach $href (@demos) {
 	my $do_next = 0;
 
 	# parent detected
-	if (defined @parents) {
+	if (@parents) {
 	    foreach $foo (@parents) {
 		if ($foo eq $parent_name) {
 		    $do_next = 1;
@@ -54,7 +54,7 @@ foreach $href (@demos) {
 
 	push @parents, $parent_name;
 
-	$tmp = (defined @child_arrays)?($#child_arrays + 1):0;
+	$tmp = @child_arrays?($#child_arrays + 1):0;
 	push @child_arrays, "child$tmp";
 
 	push @demos, {"title" => $parent_name, "file" => "NULL",
@@ -62,7 +62,7 @@ foreach $href (@demos) {
     }
 }
 
-if (defined @parents) {
+if (@parents) {
     $i = 0;
     for ($i = 0; $i <= $#parents; $i++) {
 	$first = 1;
@@ -105,7 +105,7 @@ if (defined @parents) {
 } @demos_old;
 
 # sort the child arrays
-if (defined @child_arrays) {
+if (@child_arrays) {
     for ($i = 0; $i <= $#child_arrays; $i++) {
 	@foo_old = @{$child_arrays[$i]};
 
@@ -133,7 +133,7 @@ foreach $href (@demos) {
 	print ", \n";
     }
 
-    if (defined @parents) {
+    if (@parents) {
 	for ($i = 0; $i <= $#parents; $i++) {
 	    if ($parents[$i] eq $href->{title}) {
 
diff -udpr gtk+-2.6.7_/gdk/win32/gdkcursor-win32.c gtk+-2.6.7/gdk/win32/gdkcursor-win32.c
--- gtk+-2.6.7_/gdk/win32/gdkcursor-win32.c	2004-03-06 05:37:07.000000000 +0200
+++ gtk+-2.6.7/gdk/win32/gdkcursor-win32.c	2020-12-27 16:22:52.529875169 +0200
@@ -41,6 +41,12 @@ _gdk_win32_data_to_wcursor (GdkCursorTyp
   if (i >= G_N_ELEMENTS (cursors) || !cursors[i].name)
     return NULL;
 
+  /* use real win32 cursor if possible */
+  if (cursors[i].builtin)
+    {
+      return LoadCursor (NULL, cursors[i].builtin);
+    }
+
   w = GetSystemMetrics (SM_CXCURSOR);
   h = GetSystemMetrics (SM_CYCURSOR);
 
diff -udpr gtk+-2.6.7_/gdk/win32/gdkdnd-win32.c gtk+-2.6.7/gdk/win32/gdkdnd-win32.c
--- gtk+-2.6.7_/gdk/win32/gdkdnd-win32.c	2004-12-19 23:00:58.000000000 +0200
+++ gtk+-2.6.7/gdk/win32/gdkdnd-win32.c	2020-12-27 16:22:52.529875169 +0200
@@ -978,6 +978,8 @@ gdk_dropfiles_filter (GdkXEvent *xev,
       /* WM_DROPFILES drops are always file names */
       context->targets =
 	g_list_append (NULL, GUINT_TO_POINTER (_text_uri_list));
+      context->actions = GDK_ACTION_COPY;
+      context->suggested_action = GDK_ACTION_COPY;
       current_dest_drag = context;
 
       event->dnd.type = GDK_DROP_START;
@@ -987,8 +989,8 @@ gdk_dropfiles_filter (GdkXEvent *xev,
       DragQueryPoint (hdrop, &pt);
       ClientToScreen (msg->hwnd, &pt);
 
-      event->dnd.x_root = pt.x;
-      event->dnd.y_root = pt.y;
+      event->dnd.x_root = pt.x + _gdk_offset_x;
+      event->dnd.y_root = pt.y + _gdk_offset_y;
       event->dnd.time = _gdk_win32_get_next_tick (msg->time);
 
       nfiles = DragQueryFile (hdrop, 0xFFFFFFFF, NULL, 0);
diff -udpr gtk+-2.6.7_/gdk/win32/gdkdrawable-win32.c gtk+-2.6.7/gdk/win32/gdkdrawable-win32.c
--- gtk+-2.6.7_/gdk/win32/gdkdrawable-win32.c	2005-04-03 23:36:39.000000000 +0300
+++ gtk+-2.6.7/gdk/win32/gdkdrawable-win32.c	2020-12-27 16:22:52.530875169 +0200
@@ -1856,6 +1856,16 @@ _gdk_win32_blit (gboolean              u
 			   xdest, ydest,
 			   use_fg_bg));
 
+  /* If blitting from the root window, take the multi-monitor offset
+   * into account.
+   */
+  if (src == ((GdkWindowObject *)_gdk_parent_root)->impl)
+    {
+      GDK_NOTE (MISC, g_print ("... offsetting src coords\n"));
+      xsrc -= _gdk_offset_x;
+      ysrc -= _gdk_offset_y;
+    }
+
   draw_impl = (GdkDrawableImplWin32 *) drawable;
 
   if (GDK_IS_DRAWABLE_IMPL_WIN32 (src))
diff -udpr gtk+-2.6.7_/gdk/win32/gdkevents-win32.c gtk+-2.6.7/gdk/win32/gdkevents-win32.c
--- gtk+-2.6.7_/gdk/win32/gdkevents-win32.c	2005-04-04 02:40:42.000000000 +0300
+++ gtk+-2.6.7/gdk/win32/gdkevents-win32.c	2020-12-27 16:22:52.530875169 +0200
@@ -162,7 +162,8 @@ static HKL latin_locale = NULL;
 #endif
 
 static gboolean in_ime_composition = FALSE;
-static UINT     resize_timer;
+static UINT     modal_timer;
+static UINT     sync_timer = 0;
 
 static int debug_indent = 0;
 
@@ -1841,140 +1842,7 @@ handle_configure_event (MSG       *msg,
     }
 }
 
-static void
-erase_background (GdkWindow *window,
-		  HDC        hdc)
-{
-  HDC bgdc = NULL;
-  HBRUSH hbr = NULL;
-  HPALETTE holdpal = NULL;
-  RECT rect;
-  COLORREF bg;
-  GdkColormap *colormap;
-  GdkColormapPrivateWin32 *colormap_private;
-  int x, y;
-  int x_offset, y_offset;
-  
-  if (((GdkWindowObject *) window)->input_only ||
-      ((GdkWindowObject *) window)->bg_pixmap == GDK_NO_BG ||
-      GDK_WINDOW_IMPL_WIN32 (((GdkWindowObject *) window)->impl)->position_info.no_bg)
-    {
-      return;
-    }
-
-  colormap = gdk_drawable_get_colormap (window);
-
-  if (colormap &&
-      (colormap->visual->type == GDK_VISUAL_PSEUDO_COLOR ||
-       colormap->visual->type == GDK_VISUAL_STATIC_COLOR))
-    {
-      int k;
-	  
-      colormap_private = GDK_WIN32_COLORMAP_DATA (colormap);
-
-      if (!(holdpal = SelectPalette (hdc,  colormap_private->hpal, FALSE)))
-        WIN32_GDI_FAILED ("SelectPalette");
-      else if ((k = RealizePalette (hdc)) == GDI_ERROR)
-	WIN32_GDI_FAILED ("RealizePalette");
-      else if (k > 0)
-	GDK_NOTE (COLORMAP, g_print ("erase_background: realized %p: %d colors\n",
-				     colormap_private->hpal, k));
-    }
-  
-  x_offset = y_offset = 0;
-  while (window && ((GdkWindowObject *) window)->bg_pixmap == GDK_PARENT_RELATIVE_BG)
-    {
-      /* If this window should have the same background as the parent,
-       * fetch the parent. (And if the same goes for the parent, fetch
-       * the grandparent, etc.)
-       */
-      x_offset += ((GdkWindowObject *) window)->x;
-      y_offset += ((GdkWindowObject *) window)->y;
-      window = GDK_WINDOW (((GdkWindowObject *) window)->parent);
-    }
-  
-  if (GDK_WINDOW_IMPL_WIN32 (((GdkWindowObject *) window)->impl)->position_info.no_bg)
-    {
-      /* Improves scolling effect, e.g. main buttons of testgtk */
-      return;
-    }
-
-  GetClipBox (hdc, &rect);
-
-  if (((GdkWindowObject *) window)->bg_pixmap == NULL)
-    {
-      bg = _gdk_win32_colormap_color (GDK_DRAWABLE_IMPL_WIN32 (((GdkWindowObject *) window)->impl)->colormap,
-				      ((GdkWindowObject *) window)->bg_color.pixel);
-      
-      if (!(hbr = CreateSolidBrush (bg)))
-	WIN32_GDI_FAILED ("CreateSolidBrush");
-      else if (!FillRect (hdc, &rect, hbr))
-	WIN32_GDI_FAILED ("FillRect");
-      if (hbr != NULL)
-	DeleteObject (hbr);
-    }
-  else if (((GdkWindowObject *) window)->bg_pixmap != GDK_NO_BG)
-    {
-      GdkPixmap *pixmap = ((GdkWindowObject *) window)->bg_pixmap;
-      GdkPixmapImplWin32 *pixmap_impl = GDK_PIXMAP_IMPL_WIN32 (GDK_PIXMAP_OBJECT (pixmap)->impl);
-      
-      if (x_offset == 0 && y_offset == 0 &&
-	  pixmap_impl->width <= 8 && pixmap_impl->height <= 8)
-	{
-	  if (!(hbr = CreatePatternBrush (GDK_PIXMAP_HBITMAP (pixmap))))
-	    WIN32_GDI_FAILED ("CreatePatternBrush");
-	  else if (!FillRect (hdc, &rect, hbr))
-	    WIN32_GDI_FAILED ("FillRect");
-	  if (hbr != NULL)
-	    DeleteObject (hbr);
-	}
-      else
-	{
-	  HGDIOBJ oldbitmap;
-
-	  if (!(bgdc = CreateCompatibleDC (hdc)))
-	    {
-	      WIN32_GDI_FAILED ("CreateCompatibleDC");
-	      return;
-	    }
-	  if (!(oldbitmap = SelectObject (bgdc, GDK_PIXMAP_HBITMAP (pixmap))))
-	    {
-	      WIN32_GDI_FAILED ("SelectObject");
-	      DeleteDC (bgdc);
-	      return;
-	    }
-	  x = -x_offset;
-	  while (x < rect.right)
-	    {
-	      if (x + pixmap_impl->width >= rect.left)
-		{
-		  y = -y_offset;
-		  while (y < rect.bottom)
-		    {
-		      if (y + pixmap_impl->height >= rect.top)
-			{
-			  if (!BitBlt (hdc, x, y,
-				       pixmap_impl->width, pixmap_impl->height,
-				       bgdc, 0, 0, SRCCOPY))
-			    {
-			      WIN32_GDI_FAILED ("BitBlt");
-			      SelectObject (bgdc, oldbitmap);
-			      DeleteDC (bgdc);
-			      return;
-			    }
-			}
-		      y += pixmap_impl->height;
-		    }
-		}
-	      x += pixmap_impl->width;
-	    }
-	  SelectObject (bgdc, oldbitmap);
-	  DeleteDC (bgdc);
-	}
-    }
-}
-
-static GdkRegion *
+GdkRegion *
 _gdk_win32_hrgn_to_region (HRGN hrgn)
 {
   RGNDATA *rgndata;
@@ -2043,6 +1911,7 @@ handle_wm_paint (MSG        *msg,
   if (GetUpdateRgn (msg->hwnd, hrgn, FALSE) == ERROR)
     {
       WIN32_GDI_FAILED ("GetUpdateRgn");
+      DeleteObject (hrgn);
       return;
     }
 
@@ -2115,6 +1984,7 @@ handle_wm_paint (MSG        *msg,
 	    }
 	}
 
+      DeleteObject (hrgn);
       return;
     }
 
@@ -2138,15 +2008,32 @@ handle_stuff_while_moving_or_resizing (v
 }
 
 static VOID CALLBACK
-resize_timer_proc (HWND     hwnd,
-		   UINT     msg,
-		   UINT     id,
-		   DWORD    time)
+modal_timer_proc (HWND     hwnd,
+		  UINT     msg,
+		  UINT     id,
+		  DWORD    time)
 {
   if (_sizemove_in_progress)
     handle_stuff_while_moving_or_resizing ();
 }
 
+static VOID CALLBACK
+sync_timer_proc (HWND hwnd,
+		 UINT msg,
+		 UINT id,
+		 DWORD time)
+{
+  MSG message;
+  if (PeekMessageW (&message, hwnd, WM_PAINT, WM_PAINT, PM_REMOVE))
+    {
+      return;
+    }
+
+  RedrawWindow (hwnd, NULL, NULL, RDW_INVALIDATE|RDW_UPDATENOW|RDW_ALLCHILDREN);
+
+  KillTimer (hwnd, sync_timer);
+}
+
 static void
 handle_display_change (void)
 {
@@ -2923,11 +2810,17 @@ gdk_event_translate (GdkDisplay *display
       if (GDK_WINDOW_DESTROYED (window))
 	break;
 
-      erase_background (window, (HDC) msg->wParam);
       return_val = TRUE;
       *ret_valp = 1;
       break;
 
+    case WM_SYNCPAINT:
+
+      sync_timer = SetTimer (GDK_WINDOW_HWND (window),
+			     1,
+			     200, sync_timer_proc);
+      break;
+
     case WM_PAINT:
       handle_wm_paint (msg, window, FALSE, NULL);
       break;
@@ -3042,15 +2935,31 @@ gdk_event_translate (GdkDisplay *display
 
     case WM_ENTERSIZEMOVE:
       _sizemove_in_progress = TRUE;
-      resize_timer = SetTimer (NULL, 0, 20, resize_timer_proc);
+      modal_timer = SetTimer (NULL, 0, 20, modal_timer_proc);
       break;
 
     case WM_EXITSIZEMOVE:
       _sizemove_in_progress = FALSE;
-      KillTimer (NULL, resize_timer);
+      KillTimer (NULL, modal_timer);
+      break;
+
+    case WM_ENTERMENULOOP:
+      _sizemove_in_progress = TRUE;
+      modal_timer = SetTimer (NULL, 0, 20, modal_timer_proc);
+      break;
+
+    case WM_EXITMENULOOP:
+      _sizemove_in_progress = FALSE;
+      KillTimer (NULL, modal_timer);
       break;
 
     case WM_WINDOWPOSCHANGED :
+       /* If position and size haven't changed, don't do anything */
+      if (_sizemove_in_progress &&
+ 	  (((WINDOWPOS *)msg->lParam)->flags & SWP_NOMOVE) &&
+ 	  (((WINDOWPOS *)msg->lParam)->flags & SWP_NOSIZE))
+ 	break;
+ 
       /* Once we've entered the moving or sizing modal loop, we won't
        * return to the main loop until we're done sizing or moving.
        */
@@ -3058,55 +2967,14 @@ gdk_event_translate (GdkDisplay *display
 	 GDK_WINDOW_TYPE (window) != GDK_WINDOW_CHILD &&
 	 !GDK_WINDOW_DESTROYED (window))
 	{
-	  RECT client_rect;
-	  POINT point;
-
-	  GetClientRect (msg->hwnd, &client_rect);
-	  point.x = client_rect.left; /* always 0 */
-	  point.y = client_rect.top;
-	  /* top level windows need screen coords */
-	  if (gdk_window_get_parent (window) == _gdk_parent_root)
-	    {
-	      ClientToScreen (msg->hwnd, &point);
-	      point.x += _gdk_offset_x;
-	      point.y += _gdk_offset_y;
-	    }
-  
-	  GDK_WINDOW_IMPL_WIN32 (((GdkWindowObject *) window)->impl)->width = client_rect.right - client_rect.left;
-	  GDK_WINDOW_IMPL_WIN32 (((GdkWindowObject *) window)->impl)->height = client_rect.bottom - client_rect.top;
-	  
-	  ((GdkWindowObject *) window)->x = point.x;
-	  ((GdkWindowObject *) window)->y = point.y;
-
 	  if (((GdkWindowObject *) window)->event_mask & GDK_STRUCTURE_MASK)
 	    {
-	      GdkEvent *event = gdk_event_new (GDK_CONFIGURE);
-	      
-	      event->configure.window = window;
-	      
-	      event->configure.width = client_rect.right - client_rect.left;
-	      event->configure.height = client_rect.bottom - client_rect.top;
-	      
-	      event->configure.x = point.x;
-	      event->configure.y = point.y;
-	      
 	      if (((GdkWindowObject *) window)->resize_count > 1)
 		((GdkWindowObject *) window)->resize_count -= 1;
 	      
-#if 0 /* I don't like calling _gdk_event_func() from here, isn't there
-       * a risk of getting events processed in the wrong order, like
-       * Owen says in the discussion of bug #99540?
-       */
-	      fixup_event (event);
-	      if (_gdk_event_func)
-		(*_gdk_event_func) (event, _gdk_event_data);
-	      gdk_event_free (event);
-#else /* Calling append_event() is slower, but guarantees that events won't
-       * get reordered, I think.
-       */
-	      append_event (display, event);
-#endif
-	      
+ 	      handle_configure_event (msg, window);
+	      g_main_context_iteration (NULL, FALSE);
+
 	      /* Dispatch main loop - to realize resizes... */
 	      handle_stuff_while_moving_or_resizing ();
 	      
@@ -3119,6 +2987,7 @@ gdk_event_translate (GdkDisplay *display
 
     case WM_SIZING:
       GetWindowRect (GDK_WINDOW_HWND (window), &rect);
+      drag = (RECT *) msg->lParam;
       GDK_NOTE (EVENTS, g_print (" %s curr:%s drag:%s",
 				 (msg->wParam == WMSZ_BOTTOM ? "BOTTOM" :
 				  (msg->wParam == WMSZ_BOTTOMLEFT ? "BOTTOMLEFT" :
@@ -3131,13 +3000,13 @@ gdk_event_translate (GdkDisplay *display
 					(msg->wParam == WMSZ_BOTTOMRIGHT ? "BOTTOMRIGHT" :
 					 "???")))))))),
 				 _gdk_win32_rect_to_string (&rect),
-				 _gdk_win32_rect_to_string ((RECT *) msg->lParam)));
+				 _gdk_win32_rect_to_string (drag)));
 
       impl = GDK_WINDOW_IMPL_WIN32 (((GdkWindowObject *) window)->impl);
-      drag = (RECT *) msg->lParam;
       orig_drag = *drag;
       if (impl->hint_flags & GDK_HINT_RESIZE_INC)
 	{
+	  GDK_NOTE (EVENTS, g_print (" (RESIZE_INC)"));
 	  if (impl->hint_flags & GDK_HINT_BASE_SIZE)
 	    {
 	      /* Resize in increments relative to the base size */
@@ -3166,7 +3035,6 @@ gdk_event_translate (GdkDisplay *display
 	      if (drag->bottom == rect.bottom)
 		break;
 	      adjust_drag (&drag->bottom, rect.bottom, impl->hints.height_inc);
-
 	      break;
 
 	    case WMSZ_BOTTOMLEFT:
@@ -3221,7 +3089,7 @@ gdk_event_translate (GdkDisplay *display
 	    {
 	      *ret_valp = TRUE;
 	      return_val = TRUE;
-	      GDK_NOTE (EVENTS, g_print (" (handled RESIZE_INC: drag:%s)",
+	      GDK_NOTE (EVENTS, g_print (" (handled RESIZE_INC: %s)",
 					 _gdk_win32_rect_to_string (drag)));
 	    }
 	}
@@ -3230,18 +3098,105 @@ gdk_event_translate (GdkDisplay *display
 
       if (impl->hint_flags & GDK_HINT_ASPECT)
 	{
-	  gdouble drag_aspect = (gdouble) (drag->right - drag->left) / (drag->bottom - drag->top);
-	  
-	  GDK_NOTE (EVENTS, g_print (" (aspect:%g)", drag_aspect));
-	  if (drag_aspect < impl->hints.min_aspect ||
-	      drag_aspect > impl->hints.max_aspect)
+	  RECT decorated_rect;
+	  RECT undecorated_drag;
+	  int decoration_width, decoration_height;
+	  gdouble drag_aspect;
+	  int drag_width, drag_height, new_width, new_height;
+
+	  GetClientRect (GDK_WINDOW_HWND (window), &rect);
+	  decorated_rect = rect;
+	  _gdk_win32_adjust_client_rect (window, &decorated_rect);
+
+	  /* Set undecorated_drag to the client area being dragged
+	   * out, in screen coordinates.
+	   */
+	  undecorated_drag = *drag;
+	  undecorated_drag.left -= decorated_rect.left - rect.left;
+	  undecorated_drag.right -= decorated_rect.right - rect.right;
+	  undecorated_drag.top -= decorated_rect.top - rect.top;
+	  undecorated_drag.bottom -= decorated_rect.bottom - rect.bottom;
+
+	  decoration_width = (decorated_rect.right - decorated_rect.left) - (rect.right - rect.left);
+	  decoration_height = (decorated_rect.bottom - decorated_rect.top) - (rect.bottom - rect.top);
+
+	  drag_width = undecorated_drag.right - undecorated_drag.left;
+	  drag_height = undecorated_drag.bottom - undecorated_drag.top;
+
+	  drag_aspect = (gdouble) drag_width / drag_height;
+
+	  GDK_NOTE (EVENTS, g_print (" (ASPECT:%g--%g curr: %g)",
+				     impl->hints.min_aspect, impl->hints.max_aspect, drag_aspect));
+
+	  if (drag_aspect < impl->hints.min_aspect)
 	    {
-	      *drag = rect;
-	      *ret_valp = TRUE;
-	      return_val = TRUE;
-	      GDK_NOTE (EVENTS, g_print (" (handled ASPECT: drag:%s)",
-					 _gdk_win32_rect_to_string (drag)));
+	      /* Aspect is getting too narrow */
+	      switch (msg->wParam)
+		{
+		case WMSZ_BOTTOM:
+		case WMSZ_TOP:
+		  /* User drags top or bottom edge outward. Keep height, increase width. */
+		  new_width = impl->hints.min_aspect * drag_height;
+		  drag->left -= (new_width - drag_width) / 2;
+		  drag->right = drag->left + new_width + decoration_width;
+		  break;
+		case WMSZ_BOTTOMLEFT:
+		case WMSZ_BOTTOMRIGHT:
+		  /* User drags bottom-left or bottom-right corner down. Adjust height. */
+		  new_height = drag_width / impl->hints.min_aspect;
+		  drag->bottom = drag->top + new_height + decoration_height;
+		  break;
+		case WMSZ_LEFT:
+		case WMSZ_RIGHT:
+		  /* User drags left or right edge inward. Decrease height */
+		  new_height = drag_width / impl->hints.min_aspect;
+		  drag->top += (drag_height - new_height) / 2;
+		  drag->bottom = drag->top + new_height + decoration_height;
+		  break;
+		case WMSZ_TOPLEFT:
+		case WMSZ_TOPRIGHT:
+		  /* User drags top-left or top-right corner up. Adjust height. */
+		  new_height = drag_width / impl->hints.min_aspect;
+		  drag->top = drag->bottom - new_height - decoration_height;
+		}
+	    }
+	  else if (drag_aspect > impl->hints.max_aspect)
+	    {
+	      /* Aspect is getting too wide */
+	      switch (msg->wParam)
+		{
+		case WMSZ_BOTTOM:
+		case WMSZ_TOP:
+		  /* User drags top or bottom edge inward. Decrease width. */
+		  new_width = impl->hints.max_aspect * drag_height;
+		  drag->left += (drag_width - new_width) / 2;
+		  drag->right = drag->left + new_width + decoration_width;
+		  break;
+		case WMSZ_BOTTOMLEFT:
+		case WMSZ_TOPLEFT:
+		  /* User drags bottom-left or top-left corner left. Adjust width. */
+		  new_width = impl->hints.max_aspect * drag_height;
+		  drag->left = drag->right - new_width - decoration_width;
+		  break;
+		case WMSZ_BOTTOMRIGHT:
+		case WMSZ_TOPRIGHT:
+		  /* User drags bottom-right or top-right corner right. Adjust width. */
+		  new_width = impl->hints.max_aspect * drag_height;
+		  drag->right = drag->left + new_width + decoration_width;
+		  break;
+		case WMSZ_LEFT:
+		case WMSZ_RIGHT:
+		  /* User drags left or right edge outward. Increase height. */
+		  new_height = drag_width / impl->hints.max_aspect;
+		  drag->top -= (new_height - drag_height) / 2;
+		  drag->bottom = drag->top + new_height + decoration_height;
+		  break;
+		}
 	    }
+	  *ret_valp = TRUE;
+	  return_val = TRUE;
+	  GDK_NOTE (EVENTS, g_print (" (handled ASPECT: %s)",
+				     _gdk_win32_rect_to_string (drag)));
 	}
       break;
 
diff -udpr gtk+-2.6.7_/gdk/win32/gdkgeometry-win32.c gtk+-2.6.7/gdk/win32/gdkgeometry-win32.c
--- gtk+-2.6.7_/gdk/win32/gdkgeometry-win32.c	2005-03-13 00:07:34.000000000 +0200
+++ gtk+-2.6.7/gdk/win32/gdkgeometry-win32.c	2020-12-27 16:22:52.531875169 +0200
@@ -108,9 +108,9 @@ gdk_window_scroll (GdkWindow *window,
   GdkRegion *invalidate_region;
   GdkWindowImplWin32 *impl;
   GdkWindowObject *obj;
-  GdkRectangle dest_rect;
   GList *tmp_list;
   GdkWindowParentPos parent_pos;
+  HRGN native_invalidate_region;
   
   g_return_if_fail (GDK_IS_WINDOW (window));
 
@@ -130,22 +130,6 @@ gdk_window_scroll (GdkWindow *window,
   if (obj->update_area)
     gdk_region_offset (obj->update_area, dx, dy);
   
-  invalidate_region = gdk_region_rectangle (&impl->position_info.clip_rect);
-  
-  dest_rect = impl->position_info.clip_rect;
-  dest_rect.x += dx;
-  dest_rect.y += dy;
-  gdk_rectangle_intersect (&dest_rect, &impl->position_info.clip_rect, &dest_rect);
-
-  if (dest_rect.width > 0 && dest_rect.height > 0)
-    {
-      GdkRegion *tmp_region;
-
-      tmp_region = gdk_region_rectangle (&dest_rect);
-      gdk_region_subtract (invalidate_region, tmp_region);
-      gdk_region_destroy (tmp_region);
-    }
-  
   gdk_window_compute_parent_pos (impl, &parent_pos);
 
   parent_pos.x += obj->x;
@@ -156,10 +140,13 @@ gdk_window_scroll (GdkWindow *window,
 
   gdk_window_tmp_unset_bg (window);
 
-  if (!ScrollWindowEx (GDK_WINDOW_HWND (window),
-		       dx, dy, NULL, NULL,
-		       NULL, NULL, SW_SCROLLCHILDREN))
-    WIN32_API_FAILED ("ScrollWindowEx");
+  native_invalidate_region = CreateRectRgn (0, 0, 0, 0);
+  if (native_invalidate_region == NULL)
+    WIN32_API_FAILED ("CreateRectRgn");
+
+  API_CALL (ScrollWindowEx, (GDK_WINDOW_HWND (window),
+			     dx, dy, NULL, NULL,
+			     native_invalidate_region, NULL, SW_SCROLLCHILDREN));
 
   if (impl->position_info.no_bg)
     gdk_window_tmp_reset_bg (window);
@@ -173,8 +160,15 @@ gdk_window_scroll (GdkWindow *window,
       tmp_list = tmp_list->next;
     }
 
-  gdk_window_invalidate_region (window, invalidate_region, TRUE);
-  gdk_region_destroy (invalidate_region);
+  if (native_invalidate_region != NULL)
+    {
+      invalidate_region = _gdk_win32_hrgn_to_region (native_invalidate_region);
+      gdk_region_offset (invalidate_region, impl->position_info.x_offset,
+                         impl->position_info.y_offset);
+      gdk_window_invalidate_region (window, invalidate_region, TRUE);
+      gdk_region_destroy (invalidate_region);
+      GDI_CALL (DeleteObject, (native_invalidate_region));
+    }
 }
 
 void
diff -udpr gtk+-2.6.7_/gdk/win32/gdkinput-win32.c gtk+-2.6.7/gdk/win32/gdkinput-win32.c
--- gtk+-2.6.7_/gdk/win32/gdkinput-win32.c	2005-02-25 02:09:52.000000000 +0200
+++ gtk+-2.6.7/gdk/win32/gdkinput-win32.c	2020-12-27 16:22:52.531875169 +0200
@@ -43,16 +43,10 @@
 #define PACKETMODE (PK_BUTTONS)
 #include <pktdef.h>
 
-/* If USE_SYSCONTEXT is on, we open the Wintab device (hmm, what if
- * there are several?) as a system pointing device, i.e. it controls
- * the normal Windows cursor. This seems much more natural.
- */
-#define USE_SYSCONTEXT 1	/* The code for the other choice is not
-				 * good at all.
-				 */
-
 #define DEBUG_WINTAB 1		/* Verbose debug messages enabled */
 
+#define PROXIMITY_OUT_DELAY 200 /* In milliseconds, see set_ignore_core */
+
 #endif
 
 #if defined(HAVE_WINTAB) || defined(HAVE_WHATEVER_OTHER)
@@ -63,10 +57,6 @@
 
 /* Forward declarations */
 
-#if !USE_SYSCONTEXT
-static GdkInputWindow *gdk_input_window_find_within (GdkWindow *window);
-#endif
-
 #ifdef HAVE_WINTAB
 
 static GdkDevicePrivate *gdk_input_find_dev_from_ctx (HCTX hctx,
@@ -211,7 +201,6 @@ _gdk_input_wintab_init_check (void)
   GdkDevicePrivate *gdkdev;
   GdkWindowAttr wa;
   WORD specversion;
-  LOGCONTEXT defcontext;
   HCTX *hctx;
   UINT ndevices, ncursors, ncsrtypes, firstcsr, hardware;
   BOOL active;
@@ -219,6 +208,7 @@ _gdk_input_wintab_init_check (void)
   int i, k;
   int devix, cursorix;
   char devname[100], csrname[100];
+  BOOL defcontext_done;
 
   if (wintab_initialized)
     return;
@@ -233,17 +223,6 @@ _gdk_input_wintab_init_check (void)
       WTInfo (WTI_INTERFACE, IFC_SPECVERSION, &specversion);
       GDK_NOTE (INPUT, g_print ("Wintab interface version %d.%d\n",
 			       HIBYTE (specversion), LOBYTE (specversion)));
-#if USE_SYSCONTEXT
-      WTInfo (WTI_DEFSYSCTX, 0, &defcontext);
-#if DEBUG_WINTAB
-      GDK_NOTE (INPUT, (g_print("DEFSYSCTX:\n"), print_lc(&defcontext)));
-#endif
-#else
-      WTInfo (WTI_DEFCONTEXT, 0, &defcontext);
-#if DEBUG_WINTAB
-      GDK_NOTE (INPUT, (g_print("DEFCONTEXT:\n"), print_lc(&defcontext)));
-#endif
-#endif
       WTInfo (WTI_INTERFACE, IFC_NDEVICES, &ndevices);
       WTInfo (WTI_INTERFACE, IFC_NCURSORS, &ncursors);
 #if DEBUG_WINTAB
@@ -268,6 +247,11 @@ _gdk_input_wintab_init_check (void)
       for (devix = 0; devix < ndevices; devix++)
 	{
 	  LOGCONTEXT lc;
+	  
+	  /* We open the Wintab device (hmm, what if there are several?) as a
+	   * system pointing device, i.e. it controls the normal Windows
+	   * cursor. This seems much more natural.
+	   */
 
 	  WTInfo (WTI_DEVICES + devix, DVC_NAME, devname);
       
@@ -279,65 +263,39 @@ _gdk_input_wintab_init_check (void)
 	  WTInfo (WTI_DEVICES + devix, DVC_NPRESSURE, &axis_npressure);
 	  WTInfo (WTI_DEVICES + devix, DVC_ORIENTATION, axis_or);
 
+	  defcontext_done = FALSE;
 	  if (HIBYTE (specversion) > 1 || LOBYTE (specversion) >= 1)
 	    {
-	      WTInfo (WTI_DDCTXS + devix, CTX_NAME, lc.lcName);
-	      WTInfo (WTI_DDCTXS + devix, CTX_OPTIONS, &lc.lcOptions);
-	      lc.lcOptions |= CXO_MESSAGES;
-#if USE_SYSCONTEXT
-	      lc.lcOptions |= CXO_SYSTEM;
+	      /* Try to get device-specific default context */
+	      /* Some drivers, e.g. Aiptek, don't provide this info */
+	      if (WTInfo (WTI_DSCTXS + devix, 0, &lc) > 0)
+		defcontext_done = TRUE;
+#if DEBUG_WINTAB
+	      if (defcontext_done)
+		GDK_NOTE (INPUT, (g_print("Using device-specific default context\n")));
+	      else
+		GDK_NOTE (INPUT, (g_print("Note: Driver did not provide device specific default context info despite claiming to support version 1.1\n")));
 #endif
-	      lc.lcStatus = 0;
-	      WTInfo (WTI_DDCTXS + devix, CTX_LOCKS, &lc.lcLocks);
-	      lc.lcMsgBase = WT_DEFBASE;
-	      lc.lcDevice = devix;
-	      lc.lcPktRate = 50;
-	      lc.lcPktData = PACKETDATA;
-	      lc.lcPktMode = PK_BUTTONS; /* We want buttons in relative mode */
-	      lc.lcMoveMask = PACKETDATA;
-	      lc.lcBtnDnMask = lc.lcBtnUpMask = ~0;
-	      WTInfo (WTI_DDCTXS + devix, CTX_INORGX, &lc.lcInOrgX);
-	      WTInfo (WTI_DDCTXS + devix, CTX_INORGY, &lc.lcInOrgY);
-	      WTInfo (WTI_DDCTXS + devix, CTX_INORGZ, &lc.lcInOrgZ);
-	      WTInfo (WTI_DDCTXS + devix, CTX_INEXTX, &lc.lcInExtX);
-	      WTInfo (WTI_DDCTXS + devix, CTX_INEXTY, &lc.lcInExtY);
-	      WTInfo (WTI_DDCTXS + devix, CTX_INEXTZ, &lc.lcInExtZ);
-	      lc.lcOutOrgX = axis_x.axMin;
-	      lc.lcOutOrgY = axis_y.axMin;
-	      lc.lcOutExtX = axis_x.axMax - axis_x.axMin;
-	      lc.lcOutExtY = axis_y.axMax - axis_y.axMin;
-	      lc.lcOutExtY = -lc.lcOutExtY; /* We want Y growing downward */
-	      WTInfo (WTI_DDCTXS + devix, CTX_SENSX, &lc.lcSensX);
-	      WTInfo (WTI_DDCTXS + devix, CTX_SENSY, &lc.lcSensY);
-	      WTInfo (WTI_DDCTXS + devix, CTX_SENSZ, &lc.lcSensZ);
-	      WTInfo (WTI_DDCTXS + devix, CTX_SYSMODE, &lc.lcSysMode);
-	      WTInfo (WTI_DDCTXS + devix, CTX_SYSORGX, &lc.lcSysOrgX);
-	      WTInfo (WTI_DDCTXS + devix, CTX_SYSORGY, &lc.lcSysOrgY);
-	      WTInfo (WTI_DDCTXS + devix, CTX_SYSEXTX, &lc.lcSysExtX);
-	      WTInfo (WTI_DDCTXS + devix, CTX_SYSEXTY, &lc.lcSysExtY);
-	      WTInfo (WTI_DDCTXS + devix, CTX_SYSSENSX, &lc.lcSysSensX);
-	      WTInfo (WTI_DDCTXS + devix, CTX_SYSSENSY, &lc.lcSysSensY);
 	    }
-	  else
-	    {
-	      lc = defcontext;
-	      lc.lcOptions |= CXO_MESSAGES;
-	      lc.lcMsgBase = WT_DEFBASE;
-	      lc.lcPktRate = 50;
-	      lc.lcPktData = PACKETDATA;
-	      lc.lcPktMode = PACKETMODE;
-	      lc.lcMoveMask = PACKETDATA;
-	      lc.lcBtnUpMask = lc.lcBtnDnMask = ~0;
-#if 0
-	      lc.lcOutExtY = -lc.lcOutExtY; /* Y grows downward */
-#else
-	      lc.lcOutOrgX = axis_x.axMin;
-	      lc.lcOutOrgY = axis_y.axMin;
-	      lc.lcOutExtX = axis_x.axMax - axis_x.axMin;
-	      lc.lcOutExtY = axis_y.axMax - axis_y.axMin;
-	      lc.lcOutExtY = -lc.lcOutExtY; /* We want Y growing downward */
+
+	  if (!defcontext_done)
+	    WTInfo (WTI_DEFSYSCTX, 0, &lc);
+#if DEBUG_WINTAB
+	  GDK_NOTE (INPUT, (g_print("Default context:\n"), print_lc(&lc)));
 #endif
-	    }
+	  lc.lcOptions |= CXO_MESSAGES;
+	  lc.lcStatus = 0;
+	  lc.lcMsgBase = WT_DEFBASE;
+	  lc.lcPktRate = 50;
+	  lc.lcPktData = PACKETDATA;
+	  lc.lcPktMode = PACKETMODE;
+	  lc.lcMoveMask = PACKETDATA;
+	  lc.lcBtnUpMask = lc.lcBtnDnMask = ~0;
+	  lc.lcOutOrgX = axis_x.axMin;
+	  lc.lcOutOrgY = axis_y.axMin;
+	  lc.lcOutExtX = axis_x.axMax - axis_x.axMin;
+	  lc.lcOutExtY = axis_y.axMax - axis_y.axMin;
+	  lc.lcOutExtY = -lc.lcOutExtY; /* We want Y growing downward */
 #if DEBUG_WINTAB
 	  GDK_NOTE (INPUT, (g_print("context for device %d:\n", devix),
 			   print_lc(&lc)));
@@ -386,11 +344,7 @@ _gdk_input_wintab_init_check (void)
 	      gdkdev->info.name = g_strconcat (devname, " ", csrname, NULL);
 	      gdkdev->info.source = GDK_SOURCE_PEN;
 	      gdkdev->info.mode = GDK_MODE_SCREEN;
-#if USE_SYSCONTEXT
 	      gdkdev->info.has_cursor = TRUE;
-#else
-	      gdkdev->info.has_cursor = FALSE;
-#endif
 	      gdkdev->hctx = *hctx;
 	      gdkdev->cursor = cursorix;
 	      WTInfo (WTI_CURSORS + cursorix, CSR_PKTDATA, &gdkdev->pktdata);
@@ -527,32 +481,6 @@ decode_tilt (gint   *axis_data,
   axis_data[1] = sin (az) * cos (el) * 1000;
 }
 
-#if !USE_SYSCONTEXT
-
-static GdkInputWindow *
-gdk_input_window_find_within (GdkWindow *window)
-{
-  GList *list;
-  GdkWindow *tmpw;
-  GdkInputWindow *candidate = NULL;
-
-  for (list = _gdk_input_windows; list != NULL; list = list->next)
-    {
-      tmpw = ((GdkInputWindow *) (tmp_list->data))->window;
-      if (tmpw == window
-	  || IsChild (GDK_WINDOW_HWND (window), GDK_WINDOW_HWND (tmpw)))
-	{
-	  if (candidate)
-	    return NULL;		/* Multiple hits */
-	  candidate = (GdkInputWindow *) (list->data);
-	}
-    }
-
-  return candidate;
-}
-
-#endif /* USE_SYSCONTEXT */
-
 #endif /* HAVE_WINTAB */
 
 static void
@@ -726,15 +654,53 @@ get_modifier_key_state (void)
   return state;
 }
 
+#ifdef HAVE_WINTAB
+
+static guint ignore_core_timer = 0;
+
+static gboolean
+ignore_core_timefunc (gpointer data)
+{
+  /* The delay has passed */
+  _gdk_input_ignore_core = FALSE;
+  ignore_core_timer = 0;
+
+  return FALSE; /* remove timeout */
+}
+
+/*
+ * Set or unset the _gdk_input_ignore_core variable that tells GDK
+ * to ignore events for the core pointer when the tablet is in proximity
+ * The unsetting is delayed slightly so that if a tablet event arrives
+ * just after proximity out, it does not cause a core pointer event
+ * which e.g. causes GIMP to switch tools.
+ */
+static void
+set_ignore_core (gboolean ignore)
+{
+  if (ignore)
+    {
+      _gdk_input_ignore_core = TRUE;
+      /* Remove any pending clear */
+      if (ignore_core_timer)
+        {
+	  g_source_remove (ignore_core_timer);
+	  ignore_core_timer = 0;
+	}
+    }
+  else
+    if (!ignore_core_timer)
+      ignore_core_timer = g_timeout_add (PROXIMITY_OUT_DELAY,
+					 ignore_core_timefunc, NULL);
+}
+#endif /* HAVE_WINTAB */
+
 gboolean 
 _gdk_input_other_event (GdkEvent  *event,
 			MSG       *msg,
 			GdkWindow *window)
 {
 #ifdef HAVE_WINTAB
-#if !USE_SYSCONTEXT
-  GdkWindow *current_window;
-#endif
   GdkDisplay *display;
   GdkWindowObject *obj, *grab_obj;
   GdkInputWindow *input_window;
@@ -753,7 +719,6 @@ _gdk_input_other_event (GdkEvent  *event
       return FALSE;
     }
 
-#if USE_SYSCONTEXT
   window = gdk_window_at_pointer (&x, &y);
   if (window == NULL)
     window = _gdk_parent_root;
@@ -765,17 +730,6 @@ _gdk_input_other_event (GdkEvent  *event
 	    g_print ("gdk_input_win32_other_event: window=%p (%d,%d)\n",
 		     GDK_WINDOW_HWND (window), x, y));
   
-#else
-  /* ??? This code is pretty bogus */
-  current_window = gdk_win32_handle_table_lookup (GetActiveWindow ());
-  if (current_window == NULL)
-    return FALSE;
-  
-  input_window = gdk_input_window_find_within (current_window);
-  if (input_window == NULL)
-    return FALSE;
-#endif
-
   if (msg->message == WT_PACKET)
     {
       if (!WTPacket ((HCTX) msg->lParam, msg->wParam, &packet))
@@ -925,13 +879,6 @@ _gdk_input_other_event (GdkEvent  *event
 	  event->button.time = _gdk_win32_get_next_tick (msg->time);
 	  event->button.device = &gdkdev->info;
 	  
-#if 0
-#if USE_SYSCONTEXT
-	  /* Buttons 1 to 3 will come in as WM_[LMR]BUTTON{DOWN,UP} */
-	  if (event->button.button <= 3)
-	    return FALSE;
-#endif
-#endif
 	  event->button.axes = g_new(gdouble, gdkdev->info.num_axes);
 
 	  gdk_input_translate_coordinates (gdkdev, input_window,
@@ -1030,12 +977,12 @@ _gdk_input_other_event (GdkEvent  *event
       if (LOWORD (msg->lParam) == 0)
 	{
 	  event->proximity.type = GDK_PROXIMITY_OUT;
-	  _gdk_input_ignore_core = FALSE;
+	  set_ignore_core (FALSE);
 	}
       else
 	{
 	  event->proximity.type = GDK_PROXIMITY_IN;
-	  _gdk_input_ignore_core = TRUE;
+	  set_ignore_core (TRUE);
 	}
       event->proximity.time = _gdk_win32_get_next_tick (msg->time);
       event->proximity.device = &gdkdev->info;
diff -udpr gtk+-2.6.7_/gdk/win32/gdkkeys-win32.c gtk+-2.6.7/gdk/win32/gdkkeys-win32.c
--- gtk+-2.6.7_/gdk/win32/gdkkeys-win32.c	2005-02-24 00:02:59.000000000 +0200
+++ gtk+-2.6.7/gdk/win32/gdkkeys-win32.c	2020-12-27 16:22:52.531875169 +0200
@@ -265,7 +265,7 @@ reset_after_dead (guchar key_state[256])
     }
 }
 
-static gboolean
+static void
 handle_dead (guint  keysym,
 	     guint *ksymp)
 {
@@ -306,9 +306,12 @@ handle_dead (guint  keysym,
     case GDK_Greek_accentdieresis: /* 0x7ae */
       *ksymp = GDK_Greek_accentdieresis; break;
     default:
-      return FALSE;
+      /* By default use the keysym as such. This takes care of for
+       * instance the dead U+09CD (BENGALI VIRAMA) on the ekushey
+       * Bengali layout.
+       */
+      *ksymp = keysym; break;
     }
-  return TRUE;
 }
 
 static void
@@ -446,14 +449,7 @@ update_keymap (void)
 		      reset_after_dead (key_state);
 
 		      /* Use dead keysyms instead of "undead" ones */
-		      if (!handle_dead (keysym, ksymp))
-			GDK_NOTE (EVENTS,
-				  g_print ("Unhandled dead key cp:%d vk:%02x sc:%x ch:%02x wc:%04x keysym:%04x%s%s\n",
-					   _gdk_input_codepage, vk,
-					   scancode, chars[0],
-					   wcs[0], keysym,
-					   (shift&0x1 ? " shift" : ""),
-					   (shift&0x2 ? " altgr" : "")));
+		      handle_dead (keysym, ksymp);
 		    }
 		  else if (k == 0)
 		    {
diff -udpr gtk+-2.6.7_/gdk/win32/gdkpixmap-win32.c gtk+-2.6.7/gdk/win32/gdkpixmap-win32.c
--- gtk+-2.6.7_/gdk/win32/gdkpixmap-win32.c	2004-12-01 00:56:35.000000000 +0200
+++ gtk+-2.6.7/gdk/win32/gdkpixmap-win32.c	2020-12-27 16:24:11.140873949 +0200
@@ -35,6 +35,55 @@
 
 #include "gdkprivate-win32.h"
 
+static GHashTable *handle_ht = NULL;
+
+static guint
+gdi_handle_hash (HANDLE *handle)
+{
+  return (guint) *handle;
+}
+
+static gint
+gdi_handle_equal (HANDLE *a,
+		 HANDLE *b)
+{
+  return (*a == *b);
+}
+
+static void
+gdi_handle_table_insert (HANDLE  *handle,
+			       gpointer data)
+{
+  g_return_if_fail (handle != NULL);
+
+  if (!handle_ht)
+    handle_ht = g_hash_table_new ((GHashFunc) gdi_handle_hash,
+				  (GEqualFunc) gdi_handle_equal);
+
+  g_hash_table_insert (handle_ht, handle, data);
+}
+
+static void
+gdi_handle_table_remove (HANDLE handle)
+{
+  if (!handle_ht)
+    handle_ht = g_hash_table_new ((GHashFunc) gdi_handle_hash,
+				  (GEqualFunc) gdi_handle_equal);
+
+  g_hash_table_remove (handle_ht, &handle);
+}
+
+static gpointer
+gdi_handle_table_lookup (GdkNativeWindow handle)
+{
+  gpointer data = NULL;
+
+  if (handle_ht)
+    data = g_hash_table_lookup (handle_ht, &handle);
+  
+  return data;
+}
+
 static void gdk_pixmap_impl_win32_get_size   (GdkDrawable        *drawable,
 					      gint               *width,
 					      gint               *height);
@@ -111,7 +160,7 @@ gdk_pixmap_impl_win32_finalize (GObject
   if (!DeleteObject (GDK_PIXMAP_HBITMAP (wrapper)))
     WIN32_GDI_FAILED ("DeleteObject");
 
-  gdk_win32_handle_table_remove (GDK_PIXMAP_HBITMAP (wrapper));
+  gdi_handle_table_remove (GDK_PIXMAP_HBITMAP (wrapper));
 
   G_OBJECT_CLASS (parent_class)->finalize (object);
 }
@@ -306,7 +355,7 @@ gdk_pixmap_new (GdkDrawable *drawable,
   drawable_impl->handle = hbitmap;
   pixmap_impl->bits = bits;
 
-  gdk_win32_handle_table_insert (&GDK_PIXMAP_HBITMAP (pixmap), pixmap);
+  gdi_handle_table_insert (&GDK_PIXMAP_HBITMAP (pixmap), pixmap);
 
   return pixmap;
 }
@@ -478,7 +527,7 @@ gdk_pixmap_foreign_new (GdkNativeWindow
   pix_impl->height = size.cy;
   pix_impl->bits = NULL;
 
-  gdk_win32_handle_table_insert (&GDK_PIXMAP_HBITMAP (pixmap), pixmap);
+  gdi_handle_table_insert (&GDK_PIXMAP_HBITMAP (pixmap), pixmap);
 
   return pixmap;
 }
@@ -486,7 +535,7 @@ gdk_pixmap_foreign_new (GdkNativeWindow
 GdkPixmap*
 gdk_pixmap_lookup (GdkNativeWindow anid)
 {
-  return (GdkPixmap*) gdk_win32_handle_table_lookup (anid);
+  return (GdkPixmap*) gdi_handle_table_lookup (anid);
 }
 
 GdkPixmap*
diff -udpr gtk+-2.6.7_/gdk/win32/gdkprivate-win32.h gtk+-2.6.7/gdk/win32/gdkprivate-win32.h
--- gtk+-2.6.7_/gdk/win32/gdkprivate-win32.h	2005-04-04 02:40:42.000000000 +0300
+++ gtk+-2.6.7/gdk/win32/gdkprivate-win32.h	2020-12-27 16:22:52.531875169 +0200
@@ -360,6 +360,8 @@ HRGN	  _gdk_win32_gdkregion_to_hrgn  (Gd
 					 gint         x_origin,
 					 gint         y_origin);
 
+GdkRegion *_gdk_win32_hrgn_to_region    (HRGN hrgn);
+
 void	_gdk_win32_adjust_client_rect   (GdkWindow *window,
 					 RECT      *RECT);
 
diff -udpr gtk+-2.6.7_/gdk/win32/gdkproperty-win32.c gtk+-2.6.7/gdk/win32/gdkproperty-win32.c
--- gtk+-2.6.7_/gdk/win32/gdkproperty-win32.c	2005-04-04 02:40:42.000000000 +0300
+++ gtk+-2.6.7/gdk/win32/gdkproperty-win32.c	2020-12-27 16:22:52.531875169 +0200
@@ -561,7 +561,7 @@ gdk_screen_get_setting (GdkScreen   *scr
                         const gchar *name,
                         GValue      *value)
 {
-  g_return_val_if_fail (screen == gdk_screen_get_default (), FALSE);
+  g_return_val_if_fail (GDK_IS_SCREEN (screen), FALSE);
 
   /*
    * XXX : if these values get changed through the Windoze UI the
diff -udpr gtk+-2.6.7_/gdk/win32/gdkwindow-win32.c gtk+-2.6.7/gdk/win32/gdkwindow-win32.c
--- gtk+-2.6.7_/gdk/win32/gdkwindow-win32.c	2005-03-12 01:45:53.000000000 +0200
+++ gtk+-2.6.7/gdk/win32/gdkwindow-win32.c	2020-12-27 16:22:52.532875169 +0200
@@ -37,7 +37,13 @@
 #include "gdkprivate-win32.h"
 #include "gdkinput-win32.h"
 
-#if defined __MINGW32__ || (WINVER < 0x0500)
+#if defined __MINGW32__
+#include <w32api.h>
+#endif
+
+#if (defined __W32API_MAJOR_VERSION ? ((__W32API_MAJOR_VERSION < 3) || \
+	((__W32API_MAJOR_VERSION == 3) && (__W32API_MINOR_VERSION < 8))) : \
+	(WINVER < 0x0500))
 typedef struct { 
   DWORD        bV5Size; 
   LONG         bV5Width; 
diff -udpr gtk+-2.6.7_/gdk/win32/xcursors.h gtk+-2.6.7/gdk/win32/xcursors.h
--- gtk+-2.6.7_/gdk/win32/xcursors.h	2001-02-23 05:51:40.000000000 +0200
+++ gtk+-2.6.7/gdk/win32/xcursors.h	2020-12-27 16:22:52.532875169 +0200
@@ -1,357 +1,357 @@
-static const struct { const gchar *name; gint type; guchar width; guchar height; guchar hotx; guchar hoty; guchar *data; } cursors[] = {
-  { "X_cursor", 0, 16, 16, 7, 7, 
+static const struct { const gchar *name; const gchar *builtin; gint type; guchar width; guchar height; guchar hotx; guchar hoty; guchar *data; } cursors[] = {
+  { "X_cursor", NULL, 0, 16, 16, 7, 7, 
     "\125\000\000\125\152\100\001\251\152\220\006\251\152\244\032\251"
     "\032\251\152\244\006\252\252\220\001\252\252\100\000\152\251\000"
     "\000\152\251\000\001\252\252\100\006\252\252\220\032\251\152\244"
     "\152\244\032\251\152\220\006\251\152\100\001\251\125\000\000\125" }, 
-  { "arrow", 2, 16, 16, 14, 1, 
+  { "arrow", IDC_ARROW, 2, 16, 16, 14, 1, 
     "\000\000\000\025\000\000\001\151\000\000\026\251\000\001\152\244"
     "\000\026\252\244\001\152\252\220\006\252\252\220\005\126\252\100"
     "\000\032\252\100\000\152\151\000\001\251\151\000\006\244\144\000"
     "\032\220\144\000\152\100\020\000\031\000\000\000\004\000\000\000" }, 
-  { "based_arrow_down", 4, 10, 12, 4, 10, 
+  { "based_arrow_down", NULL, 4, 10, 12, 4, 10, 
     "\125\125\126\252\251\125\125\126\252\251\125\245\120\032\100\001"
     "\244\001\132\124\031\246\101\152\224\005\245\000\025\100" }, 
-  { "based_arrow_up", 6, 10, 12, 4, 10, 
+  { "based_arrow_up", NULL, 6, 10, 12, 4, 10, 
     "\000\120\000\032\100\026\251\101\232\144\025\245\100\032\100\001"
     "\244\005\132\125\152\252\225\125\125\152\252\225\125\125" }, 
-  { "boat", 8, 16, 9, 14, 4, 
+  { "boat", NULL, 8, 16, 9, 14, 4, 
     "\000\026\000\000\000\152\240\000\201\225\150\000\252\252\252\252"
     "\125\125\126\225\125\125\131\125\125\125\145\100\252\252\244\000"
     "\125\125\120\000" }, 
-  { "bogosity", 10, 15, 16, 7, 7, 
+  { "bogosity", NULL, 10, 15, 16, 7, 7, 
     "\125\105\105\125\251\031\032\225\144\144\145\101\221\221\220\126"
     "\126\126\125\252\252\252\226\145\145\146\131\221\221\231\146\106"
     "\106\145\231\131\131\226\252\252\252\125\225\225\225\006\106\106"
     "\101\131\031\031\126\244\144\152\125\121\121\125" }, 
-  { "bottom_left_corner", 12, 16, 16, 1, 14, 
+  { "bottom_left_corner", IDC_SIZENESW, 12, 16, 16, 1, 14, 
     "\125\000\000\000\151\000\000\000\151\025\000\120\151\031\001\220"
     "\151\031\006\100\151\031\031\000\151\031\144\000\151\031\220\000"
     "\151\032\125\120\151\032\252\220\151\025\125\120\151\000\000\000"
     "\151\125\125\125\152\252\252\251\152\252\252\251\125\125\125\125" }, 
-  { "bottom_right_corner", 14, 16, 16, 14, 14, 
+  { "bottom_right_corner", IDC_SIZENWSE, 14, 16, 16, 14, 14, 
     "\000\000\000\125\000\000\000\151\005\000\124\151\006\100\144\151"
     "\001\220\144\151\000\144\144\151\000\031\144\151\000\006\144\151"
     "\005\125\244\151\006\252\244\151\005\125\124\151\000\000\000\151"
     "\125\125\125\151\152\252\252\251\152\252\252\251\125\125\125\125" }, 
-  { "bottom_side", 16, 15, 16, 7, 14, 
+  { "bottom_side", IDC_SIZENS, 16, 15, 16, 7, 14, 
     "\000\005\100\000\000\031\000\000\000\144\000\000\001\220\000\000"
     "\006\100\000\000\031\000\000\120\144\024\001\221\221\220\001\226"
     "\131\000\001\231\220\000\001\251\000\000\001\220\000\125\125\125"
     "\125\252\252\252\226\252\252\252\125\125\125\125" }, 
-  { "bottom_tee", 18, 16, 12, 8, 10, 
+  { "bottom_tee", NULL, 18, 16, 12, 8, 10, 
     "\000\005\120\000\000\006\220\000\000\006\220\000\000\006\220\000"
     "\000\006\220\000\000\006\220\000\000\006\220\000\000\006\220\000"
     "\125\126\225\125\152\252\252\251\152\252\252\251\125\125\125\125" }, 
-  { "box_spiral", 20, 16, 16, 8, 8, 
+  { "box_spiral", NULL, 20, 16, 16, 8, 8, 
     "\252\252\252\251\225\125\125\125\232\252\252\251\231\125\125\131"
     "\231\252\252\231\231\225\125\231\231\232\251\231\231\231\131\231"
     "\231\231\231\231\231\232\231\231\231\225\131\231\231\252\251\231"
     "\231\125\125\231\232\252\252\231\225\125\125\131\252\252\252\251" }, 
-  { "center_ptr", 22, 12, 16, 5, 1, 
+  { "center_ptr", IDC_UPARROW, 22, 12, 16, 5, 1, 
     "\000\125\000\000\151\000\001\151\100\001\252\100\005\252\120\006"
     "\252\220\026\252\224\032\252\244\132\252\245\151\151\151\145\151"
     "\131\124\151\025\000\151\000\000\151\000\000\151\000\000\125\000" }, 
-  { "circle", 24, 16, 16, 8, 8, 
+  { "circle", NULL, 24, 16, 16, 8, 8, 
     "\000\025\124\000\001\132\245\100\005\252\252\120\026\252\252\224"
     "\032\245\132\244\132\220\006\245\152\100\001\251\152\100\001\251"
     "\152\100\001\251\152\100\001\251\132\220\006\245\032\245\132\244"
     "\026\252\252\224\005\252\252\120\001\132\245\100\000\025\124\000" }, 
-  { "clock", 26, 15, 16, 6, 3, 
+  { "clock", IDC_APPSTARTING, 26, 15, 16, 6, 3, 
     "\032\252\252\101\241\252\012\112\030\132\112\141\206\222\111\206"
     "\251\131\046\006\124\220\232\006\251\012\132\252\252\244\142\032"
     "\110\221\210\151\042\106\041\244\211\030\232\246\044\242\032\110"
     "\246\250\024\052\232\252\252\252\152\252\252\251" }, 
-  { "coffee_mug", 28, 16, 16, 7, 9, 
+  { "coffee_mug", NULL, 28, 16, 16, 7, 9, 
     "\002\252\252\000\011\125\125\200\051\125\125\151\046\125\126\231"
     "\145\252\251\131\245\125\125\131\245\125\125\131\145\125\125\131"
     "\045\125\125\131\046\226\126\231\151\231\231\231\251\232\231\231"
     "\246\231\226\231\145\125\125\131\045\125\125\131\012\252\252\240" }, 
-  { "cross", 30, 16, 16, 7, 7, 
-    "\000\025\120\000\000\031\220\000\000\031\220\000\000\031\220\000"
-    "\000\031\220\000\125\131\225\125\125\131\225\125\252\251\252\252"
-    "\125\125\125\125\252\251\252\252\000\031\220\000\000\031\220\000"
-    "\000\031\220\000\000\031\220\000\000\031\220\000\000\031\220\000" }, 
-  { "cross_reverse", 32, 16, 15, 7, 7, 
+  { "cross", IDC_CROSS, 30, 16, 15, 7, 7, 
+    "\000\031\220\000\000\031\220\000\000\031\220\000\000\031\220\000"
+    "\000\031\220\000\125\131\225\125\252\251\252\252\125\125\125\125"
+    "\252\251\252\252\125\131\225\125\000\031\220\000\000\031\220\000"
+    "\000\031\220\000\000\031\220\000\000\031\220\000" }, 
+  { "cross_reverse", NULL, 32, 16, 15, 7, 7, 
     "\044\030\220\140\211\030\221\211\142\130\226\044\030\230\230\220"
     "\006\050\242\100\125\210\211\125\252\240\052\252\000\001\000\000"
     "\252\240\052\252\125\210\211\125\006\050\242\100\030\230\230\220"
     "\142\130\226\044\211\030\221\211\044\030\220\140" }, 
-  { "crosshair", 34, 16, 16, 7, 7, 
-    "\000\005\100\000\000\006\100\000\000\006\100\000\000\006\100\000"
-    "\000\006\100\000\000\006\100\000\125\126\125\125\125\126\125\125"
-    "\252\251\252\252\000\006\100\000\000\006\100\000\000\006\100\000"
-    "\000\006\100\000\000\006\100\000\000\006\100\000\000\006\100\000" }, 
-  { "diamond_cross", 36, 16, 16, 7, 7, 
-    "\000\025\120\000\000\131\224\000\001\151\245\000\005\211\211\100"
-    "\026\011\202\120\130\011\200\224\145\131\225\145\252\250\252\251"
-    "\125\125\125\125\250\011\200\250\045\011\201\140\011\111\205\200"
-    "\002\131\226\000\000\231\230\000\000\051\240\000\000\011\200\000" }, 
-  { "dot", 38, 12, 12, 6, 6, 
+  { "crosshair", IDC_CROSS, 34, 16, 15, 7, 7, 
+    "\000\006\100\000\000\006\100\000\000\006\100\000\000\006\100\000"
+    "\000\006\100\000\000\006\100\000\125\126\125\125\252\251\252\252"
+    "\125\126\125\125\000\006\100\000\000\006\100\000\000\006\100\000"
+    "\000\006\100\000\000\006\100\000\000\006\100\000" }, 
+  { "diamond_cross", NULL, 36, 15, 15, 7, 7, 
+    "\000\031\220\000\001\246\220\000\031\231\220\001\222\141\220\031"
+    "\011\201\221\220\046\001\232\252\232\252\225\125\025\125\252\251"
+    "\252\251\220\046\001\221\220\230\031\001\222\141\220\001\231\231"
+    "\000\001\246\220\000\001\231\000\000" }, 
+  { "dot", NULL, 38, 12, 12, 6, 6, 
     "\001\125\100\025\252\124\032\252\244\132\252\245\152\252\251\152"
     "\252\251\152\252\251\152\252\251\132\252\245\032\252\244\025\252"
     "\124\001\125\100" }, 
-  { "dotbox", 40, 14, 14, 7, 6, 
+  { "dotbox", NULL, 40, 14, 14, 7, 6, 
     "\125\125\125\126\252\252\251\145\125\125\226\100\000\031\144\000"
     "\001\226\101\124\031\144\032\101\226\101\244\031\144\025\101\226"
     "\100\000\031\144\000\001\226\125\125\131\152\252\252\225\125\125"
     "\125" }, 
-  { "double_arrow", 42, 12, 16, 6, 8, 
+  { "double_arrow", IDC_SIZENS, 42, 12, 16, 6, 8, 
     "\000\125\000\001\151\100\005\252\120\026\252\224\132\151\245\151"
     "\151\151\125\151\125\000\151\000\000\151\000\125\151\125\151\151"
     "\151\132\151\245\026\252\224\005\252\120\001\151\100\000\125\000" }, 
-  { "draft_large", 44, 15, 16, 14, 0, 
-    "\000\000\000\024\000\000\005\140\000\001\132\100\000\126\250\000"
-    "\025\252\200\005\152\252\001\132\252\240\006\252\252\200\000\026"
-    "\250\000\001\146\240\000\026\032\000\001\140\150\000\026\001\200"
-    "\001\140\006\000\006\000\000\000\040\000\000\000" }, 
-  { "draft_small", 46, 15, 15, 14, 0, 
+  { "draft_large", NULL, 44, 15, 15, 14, 0, 
+    "\000\000\000\030\000\000\006\220\000\001\252\100\000\152\244\000"
+    "\032\252\200\006\252\251\001\252\252\240\005\125\252\100\000\031"
+    "\250\000\001\226\220\000\031\032\000\001\220\144\000\031\001\200"
+    "\001\220\005\000\011\000\000\000\000" }, 
+  { "draft_small", NULL, 46, 15, 15, 14, 0, 
     "\000\000\000\030\000\000\006\220\000\001\252\000\000\152\244\000"
     "\032\252\200\000\125\251\000\000\031\240\000\001\226\100\000\031"
     "\030\000\001\220\100\000\031\000\000\001\220\000\000\031\000\000"
     "\001\220\000\000\011\000\000\000\000" }, 
-  { "draped_box", 48, 14, 14, 7, 6, 
+  { "draped_box", NULL, 48, 14, 14, 7, 6, 
     "\125\125\125\126\252\252\251\140\145\220\226\032\132\111\146\220"
     "\151\226\244\121\251\145\032\105\226\121\244\131\152\105\032\226"
     "\151\006\231\141\245\244\226\006\131\011\152\252\252\225\125\125"
     "\125" }, 
-  { "exchange", 50, 16, 16, 7, 7, 
+  { "exchange", NULL, 50, 16, 16, 7, 7, 
     "\120\025\124\000\144\152\251\000\151\252\252\100\152\245\126\220"
     "\145\220\001\220\145\245\000\120\152\251\000\000\125\125\000\000"
     "\000\000\125\125\000\000\152\251\005\000\032\131\006\100\006\131"
     "\006\225\132\251\001\252\252\151\000\152\251\031\000\025\124\005" }, 
-  { "fleur", 52, 16, 16, 8, 8, 
+  { "fleur", IDC_SIZEALL, 52, 16, 16, 8, 8, 
     "\000\005\120\000\000\006\224\000\000\032\244\000\000\152\251\000"
     "\001\026\224\100\006\106\221\220\132\126\225\245\152\252\252\251"
     "\152\252\252\251\132\126\225\245\006\106\221\220\001\026\224\100"
     "\000\152\251\000\000\032\244\000\000\006\220\000\000\005\120\000" }, 
-  { "gobbler", 54, 16, 16, 14, 3, 
-    "\000\000\152\220\000\000\152\120\220\000\132\132\226\252\232\125"
-    "\252\252\252\125\252\251\132\120\252\245\132\120\151\125\132\220"
-    "\125\125\252\120\126\252\251\120\025\225\125\100\001\225\125\000"
-    "\001\220\000\000\001\220\000\000\006\251\000\000\005\125\000\000" }, 
-  { "gumby", 56, 16, 16, 2, 0, 
+  { "gobbler", NULL, 54, 16, 16, 14, 2, 
+    "\000\000\125\120\000\000\152\220\120\000\152\125\225\125\132\132"
+    "\226\252\232\125\252\252\252\120\252\251\132\120\252\245\132\120"
+    "\151\125\132\220\125\125\252\120\026\252\251\100\001\225\125\000"
+    "\001\220\000\000\001\220\000\000\005\225\000\000\006\251\000\000" }, 
+  { "gumby", NULL, 56, 16, 16, 2, 0, 
     "\012\252\000\000\122\125\200\000\244\225\140\000\251\231\230\000"
     "\244\225\130\000\244\232\231\120\252\225\132\244\132\225\132\252"
     "\005\225\130\152\000\225\130\152\000\226\131\252\000\226\130\152"
     "\000\226\130\025\002\126\126\000\011\126\125\200\012\250\252\200" }, 
-  { "hand1", 58, 13, 16, 12, 0, 
+  { "hand1", IDC_HAND, 58, 13, 16, 12, 0, 
     "\000\000\006\200\000\032\240\000\152\220\000\152\220\000\152\220"
     "\000\152\251\001\152\252\221\232\252\224\252\252\251\046\252\252"
     "\105\132\252\101\126\252\100\226\145\100\051\131\000\006\231\000"
     "\000\151\000\000" }, 
-  { "hand2", 60, 16, 16, 0, 1, 
+  { "hand2", IDC_HAND, 60, 16, 16, 0, 1, 
     "\025\125\100\000\152\252\220\000\225\125\144\000\152\251\131\000"
     "\026\125\126\100\001\251\126\100\006\125\126\100\001\251\131\220"
     "\006\125\145\144\001\245\225\131\000\132\125\144\000\031\145\220"
     "\000\006\126\100\000\001\231\000\000\000\144\000\000\000\020\000" }, 
-  { "heart", 62, 15, 14, 6, 8, 
+  { "heart", NULL, 62, 15, 14, 6, 8, 
     "\012\250\252\200\245\152\126\212\100\144\006\244\000\100\006\220"
     "\000\000\032\100\000\000\151\000\104\001\251\000\100\032\051\000"
     "\001\240\051\000\032\000\051\001\240\000\051\132\000\000\051\240"
     "\000\000\052\000\000" }, 
-  { "icon", 64, 16, 16, 8, 8, 
+  { "icon", NULL, 64, 16, 16, 8, 8, 
     "\252\252\252\252\246\146\146\146\231\231\231\232\246\146\146\146"
     "\231\125\125\232\246\125\125\146\231\125\125\232\246\125\125\146"
     "\231\125\125\232\246\125\125\146\231\125\125\232\246\125\125\146"
     "\231\231\231\232\246\146\146\146\231\231\231\232\252\252\252\252" }, 
-  { "iron_cross", 66, 16, 16, 8, 7, 
+  { "iron_cross", NULL, 66, 16, 16, 8, 7, 
     "\005\125\125\120\032\252\252\244\026\252\252\224\145\252\252\131"
     "\151\152\251\151\152\132\245\251\152\226\226\251\152\252\252\251"
     "\152\252\252\251\152\226\226\251\152\132\245\251\151\152\251\151"
     "\145\252\252\131\026\252\252\224\032\252\252\244\005\125\125\120" }, 
-  { "left_ptr", 68, 10, 16, 1, 1, 
+  { "left_ptr", IDC_ARROW, 68, 10, 16, 1, 1, 
     "\120\000\006\100\000\151\000\006\244\000\152\220\006\252\100\152"
     "\251\006\252\244\152\252\226\252\125\151\244\006\106\220\120\151"
     "\000\001\244\000\032\100\000\120" }, 
-  { "left_side", 70, 16, 15, 1, 7, 
+  { "left_side", IDC_SIZEWE, 70, 16, 15, 1, 7, 
     "\125\000\000\000\151\000\000\000\151\000\120\000\151\001\220\000"
     "\151\006\100\000\151\031\000\000\151\145\125\125\151\252\252\251"
     "\151\145\125\125\151\031\000\000\151\006\100\000\151\001\220\000"
     "\151\000\120\000\151\000\000\000\125\000\000\000" }, 
-  { "left_tee", 72, 12, 16, 1, 8, 
+  { "left_tee", NULL, 72, 12, 16, 1, 8, 
     "\125\000\000\151\000\000\151\000\000\151\000\000\151\000\000\151"
     "\000\000\151\125\125\152\252\251\152\252\251\151\125\125\151\000"
     "\000\151\000\000\151\000\000\151\000\000\151\000\000\125\000\000" }, 
-  { "leftbutton", 74, 16, 16, 8, 8, 
+  { "leftbutton", NULL, 74, 16, 16, 8, 8, 
     "\025\125\125\120\152\252\252\244\152\252\252\244\145\145\145\144"
     "\145\146\146\144\145\146\146\144\145\146\146\144\145\146\146\144"
     "\145\145\145\144\152\252\252\244\152\252\252\244\152\252\252\244"
     "\152\252\252\244\152\252\252\244\152\252\252\244\025\125\125\120" }, 
-  { "ll_angle", 76, 12, 12, 1, 10, 
+  { "ll_angle", NULL, 76, 12, 12, 1, 10, 
     "\125\000\000\151\000\000\151\000\000\151\000\000\151\000\000\151"
     "\000\000\151\000\000\151\000\000\151\125\125\152\252\251\152\252"
     "\251\125\125\125" }, 
-  { "lr_angle", 78, 12, 12, 10, 10, 
+  { "lr_angle", NULL, 78, 12, 12, 10, 10, 
     "\000\000\125\000\000\151\000\000\151\000\000\151\000\000\151\000"
     "\000\151\000\000\151\000\000\151\125\125\151\152\252\251\152\252"
     "\251\125\125\125" }, 
-  { "man", 80, 16, 16, 14, 5, 
+  { "man", NULL, 80, 16, 16, 14, 5, 
     "\001\132\224\000\006\251\252\000\005\131\225\100\220\006\100\004"
     "\144\032\220\052\031\145\145\232\006\246\152\105\001\146\145\000"
     "\000\045\144\000\000\031\220\000\000\145\144\000\001\226\131\000"
     "\006\131\226\100\026\144\146\120\152\220\032\245\252\220\032\252" }, 
-  { "middlebutton", 82, 16, 16, 8, 8, 
+  { "middlebutton", NULL, 82, 16, 16, 8, 8, 
     "\025\125\125\120\152\252\252\244\152\252\252\244\145\145\145\144"
     "\146\145\146\144\146\145\146\144\146\145\146\144\146\145\146\144"
     "\145\145\145\144\152\252\252\244\152\252\252\244\152\252\252\244"
     "\152\252\252\244\152\252\252\244\152\252\252\244\025\125\125\120" }, 
-  { "mouse", 84, 16, 16, 4, 1, 
+  { "mouse", NULL, 84, 16, 16, 4, 1, 
     "\000\125\100\000\001\124\000\000\000\152\000\000\000\045\000\000"
     "\025\151\125\100\125\132\125\120\152\252\252\225\225\125\125\145"
     "\232\132\132\151\232\132\132\151\232\132\132\151\225\125\125\151"
     "\005\125\125\051\001\125\124\000\000\225\140\000\000\052\200\000" }, 
-  { "pencil", 86, 13, 16, 11, 15, 
+  { "pencil", NULL, 86, 13, 16, 11, 15, 
     "\132\220\000\031\131\000\006\126\220\000\145\230\000\012\246\100"
     "\001\225\140\000\031\131\000\002\125\200\000\145\144\000\012\126"
     "\000\001\225\220\000\031\131\000\001\252\100\000\032\220\000\001"
     "\244\000\000\031" }, 
-  { "pirate", 88, 16, 16, 7, 12, 
+  { "pirate", IDC_NO, 88, 16, 16, 7, 12, 
     "\000\152\220\000\001\252\244\000\006\252\251\000\032\132\132\100"
     "\032\132\132\100\006\252\251\000\001\252\244\000\100\152\220\001"
     "\200\152\220\045\220\152\220\151\145\032\101\220\032\200\052\100"
     "\000\052\200\000\026\252\250\011\252\125\132\251\225\000\005\144" }, 
-  { "plus", 90, 12, 12, 5, 6, 
+  { "plus", NULL, 90, 12, 12, 5, 6, 
     "\000\125\000\000\151\000\000\151\000\000\151\000\125\151\125\152"
     "\252\251\152\252\251\125\151\125\000\151\000\000\151\000\000\151"
     "\000\000\125\000" }, 
-  { "question_arrow", 92, 11, 16, 5, 8, 
+  { "question_arrow", IDC_HELP, 92, 11, 16, 5, 8, 
     "\002\252\000\052\252\002\245\152\032\125\151\152\001\245\151\012"
     "\221\124\251\101\152\224\001\251\100\006\144\000\031\220\002\246"
     "\240\026\232\120\026\245\000\026\120\000\025\000" }, 
-  { "right_ptr", 94, 10, 16, 8, 1, 
+  { "right_ptr", NULL, 94, 10, 16, 8, 1, 
     "\000\000\120\000\031\000\006\220\001\251\000\152\220\032\251\006"
     "\252\221\252\251\152\252\225\132\251\001\246\220\151\031\006\220"
     "\121\244\000\032\100\000\120\000" }, 
-  { "right_side", 96, 16, 15, 14, 7, 
+  { "right_side", IDC_SIZEWE, 96, 16, 15, 14, 7, 
     "\000\000\000\125\000\000\000\151\000\005\000\151\000\006\100\151"
     "\000\001\220\151\000\000\144\151\125\125\131\151\152\252\252\151"
     "\125\125\131\151\000\000\144\151\000\001\220\151\000\006\100\151"
     "\000\005\000\151\000\000\000\151\000\000\000\125" }, 
-  { "right_tee", 98, 12, 16, 10, 8, 
+  { "right_tee", NULL, 98, 12, 16, 10, 8, 
     "\000\000\125\000\000\151\000\000\151\000\000\151\000\000\151\000"
     "\000\151\125\125\151\152\252\251\152\252\251\125\125\151\000\000"
     "\151\000\000\151\000\000\151\000\000\151\000\000\151\000\000\125" }, 
-  { "rightbutton", 100, 16, 16, 8, 8, 
+  { "rightbutton", NULL, 100, 16, 16, 8, 8, 
     "\025\125\125\120\152\252\252\244\152\252\252\244\145\145\145\144"
     "\146\146\145\144\146\146\145\144\146\146\145\144\146\146\145\144"
     "\145\145\145\144\152\252\252\244\152\252\252\244\152\252\252\244"
     "\152\252\252\244\152\252\252\244\152\252\252\244\025\125\125\120" }, 
-  { "rtl_logo", 102, 16, 16, 7, 7, 
+  { "rtl_logo", NULL, 102, 16, 16, 7, 7, 
     "\125\125\125\125\152\252\252\251\145\125\131\131\144\000\031\031"
     "\145\125\131\031\152\252\251\031\145\145\131\031\144\144\031\031"
     "\144\144\031\031\144\145\131\131\144\152\252\251\144\145\125\131"
     "\144\144\000\031\145\145\125\131\152\252\252\251\125\125\125\125" }, 
-  { "sailboat", 104, 16, 16, 8, 0, 
+  { "sailboat", NULL, 104, 16, 16, 8, 0, 
     "\000\000\120\000\000\000\124\000\000\001\144\000\000\005\145\000"
     "\000\006\151\000\000\026\151\000\000\032\151\100\000\132\152\100"
     "\000\152\152\100\001\152\152\120\001\252\152\220\005\252\152\220"
     "\006\252\152\225\026\252\152\245\132\251\132\200\025\125\124\000" }, 
-  { "sb_down_arrow", 106, 9, 16, 4, 15, 
+  { "sb_down_arrow", NULL, 106, 9, 16, 4, 15, 
     "\005\124\001\231\000\146\100\031\220\006\144\001\231\000\146\100"
     "\031\220\006\144\001\231\005\146\125\131\225\052\252\002\252\000"
     "\052\000\002\000" }, 
-  { "sb_h_double_arrow", 108, 15, 9, 7, 4, 
+  { "sb_h_double_arrow", IDC_SIZEWE, 108, 15, 9, 7, 4, 
     "\001\100\005\000\031\000\031\001\245\125\151\032\252\252\251\252"
     "\125\126\251\252\252\252\221\245\125\151\001\220\001\220\001\100"
     "\005\000" }, 
-  { "sb_left_arrow", 110, 16, 9, 0, 4, 
+  { "sb_left_arrow", NULL, 110, 16, 9, 0, 4, 
     "\000\120\000\000\001\220\000\000\006\225\125\125\032\252\252\252"
     "\152\225\125\125\032\252\252\252\006\225\125\125\001\220\000\000"
     "\000\120\000\000" }, 
-  { "sb_right_arrow", 112, 16, 9, 15, 4, 
+  { "sb_right_arrow", NULL, 112, 16, 9, 15, 4, 
     "\000\000\005\000\000\000\006\100\125\125\126\220\252\252\252\244"
     "\125\125\126\251\252\252\252\244\125\125\126\220\000\000\006\100"
     "\000\000\005\000" }, 
-  { "sb_up_arrow", 114, 9, 16, 4, 0, 
+  { "sb_up_arrow", NULL, 114, 9, 16, 4, 0, 
     "\000\200\000\250\000\252\200\252\250\126\145\125\231\120\146\100"
     "\031\220\006\144\001\231\000\146\100\031\220\006\144\001\231\000"
     "\146\100\025\120" }, 
-  { "sb_v_double_arrow", 116, 9, 15, 4, 7, 
+  { "sb_v_double_arrow", IDC_SIZENS, 116, 9, 15, 4, 7, 
     "\001\220\001\251\001\252\221\252\251\126\145\101\231\000\146\100"
     "\031\220\006\144\001\231\005\146\125\252\251\032\251\001\251\000"
     "\031\000" }, 
-  { "shuttle", 118, 16, 16, 11, 0, 
+  { "shuttle", NULL, 118, 16, 16, 11, 0, 
     "\000\000\006\100\000\000\032\220\000\000\152\244\000\000\251\252"
     "\000\030\251\252\000\144\251\252\001\224\251\252\001\224\251\252"
     "\001\224\251\252\001\224\251\252\006\224\251\252\032\224\251\252"
     "\152\250\251\252\025\244\145\144\000\120\032\244\000\000\006\220" }, 
-  { "sizing", 120, 16, 16, 8, 8, 
+  { "sizing", IDC_SIZENWSE, 120, 16, 16, 8, 8, 
     "\125\125\120\000\152\252\220\000\145\125\120\000\144\000\000\000"
     "\144\125\125\000\144\152\251\000\144\145\131\025\144\144\031\031"
     "\144\144\031\031\124\145\131\031\000\152\251\031\000\125\126\131"
     "\000\000\001\231\000\005\125\151\000\006\252\251\000\005\125\125" }, 
-  { "spider", 122, 16, 16, 6, 7, 
+  { "spider", NULL, 122, 16, 16, 6, 7, 
     "\030\000\002\100\006\000\011\000\002\000\010\000\001\200\044\000"
     "\100\225\140\001\220\152\220\152\050\152\222\220\006\252\251\000"
     "\006\252\250\000\050\152\226\200\220\152\220\152\100\225\140\001"
     "\001\200\040\000\002\100\030\000\006\000\011\000\030\000\002\100" }, 
-  { "spraycan", 124, 12, 16, 10, 2, 
+  { "spraycan", NULL, 124, 12, 16, 10, 2, 
     "\000\000\012\001\100\205\006\230\112\012\244\205\032\144\112\152"
     "\251\000\145\131\000\152\131\000\146\131\000\152\131\000\146\131"
     "\000\152\131\000\152\131\000\145\131\000\145\131\000\152\251\000" }, 
-  { "star", 126, 16, 16, 7, 7, 
+  { "star", NULL, 126, 16, 16, 7, 7, 
     "\000\002\000\000\000\011\200\000\000\011\200\000\000\030\220\000"
     "\000\044\140\000\000\140\044\000\001\140\045\100\132\202\012\224"
     "\240\000\000\051\132\200\012\224\005\202\011\100\006\011\202\100"
     "\030\044\140\220\030\220\030\220\032\100\006\220\031\000\001\220" }, 
-  { "target", 128, 16, 14, 7, 7, 
-    "\000\032\220\000\000\252\250\000\002\245\152\000\012\120\026\200"
-    "\051\000\001\240\244\002\000\150\240\011\200\051\240\024\120\051"
-    "\140\005\100\045\130\001\000\224\026\000\002\120\005\240\051\100"
-    "\001\132\225\000\000\025\120\000" }, 
-  { "tcross", 130, 15, 15, 7, 7, 
+  { "target", NULL, 128, 16, 15, 7, 7, 
+    "\000\025\120\000\000\132\224\000\001\252\251\000\006\240\052\100"
+    "\032\000\002\220\150\001\000\244\240\006\100\051\240\030\220\051"
+    "\240\006\100\051\150\001\000\244\032\000\002\220\006\240\052\100"
+    "\001\252\251\000\000\132\224\000\000\025\120\000" }, 
+  { "tcross", IDC_CROSS, 130, 15, 15, 7, 7, 
     "\000\005\100\000\000\031\000\000\000\144\000\000\001\220\000\000"
     "\006\100\000\000\031\000\005\125\145\125\132\252\252\251\125\126"
     "\125\124\000\031\000\000\000\144\000\000\001\220\000\000\006\100"
     "\000\000\031\000\000\000\124\000\000" }, 
-  { "top_left_arrow", 132, 16, 16, 1, 1, 
+  { "top_left_arrow", NULL, 132, 16, 16, 1, 1, 
     "\124\000\000\000\151\100\000\000\152\224\000\000\032\251\100\000"
     "\032\252\224\000\006\252\251\120\006\252\252\220\001\252\225\120"
     "\001\252\220\000\000\151\144\000\000\151\031\000\000\031\006\100"
     "\000\031\001\220\000\025\000\144\000\000\000\031\000\000\000\005" }, 
-  { "top_left_corner", 134, 16, 16, 1, 1, 
+  { "top_left_corner", IDC_SIZENWSE, 134, 16, 16, 1, 1, 
     "\125\125\125\125\152\252\252\251\152\252\252\251\151\125\125\125"
     "\151\000\000\000\151\025\125\120\151\032\252\220\151\032\125\120"
     "\151\031\220\000\151\031\144\000\151\031\031\000\151\031\006\100"
     "\151\031\001\220\151\025\000\120\151\000\000\000\125\000\000\000" }, 
-  { "top_right_corner", 136, 16, 16, 14, 1, 
+  { "top_right_corner", IDC_SIZENESW, 136, 16, 16, 14, 1, 
     "\125\125\125\125\152\252\252\251\152\252\252\251\125\125\125\151"
     "\000\000\000\151\005\125\124\151\006\252\244\151\005\125\244\151"
     "\000\006\144\151\000\031\144\151\000\144\144\151\001\220\144\151"
     "\006\100\144\151\005\000\124\151\000\000\000\151\000\000\000\125" }, 
-  { "top_side", 138, 15, 16, 7, 1, 
+  { "top_side", IDC_SIZENS, 138, 15, 16, 7, 1, 
     "\125\125\125\125\252\252\252\226\252\252\252\125\125\125\125\000"
     "\006\100\000\000\152\100\000\006\146\100\000\145\226\100\006\106"
     "\106\100\024\031\005\000\000\144\000\000\001\220\000\000\006\100"
     "\000\000\031\000\000\000\144\000\000\001\120\000" }, 
-  { "top_tee", 140, 16, 12, 8, 1, 
+  { "top_tee", NULL, 140, 16, 12, 8, 1, 
     "\125\125\125\125\152\252\252\251\152\252\252\251\125\126\225\125"
     "\000\006\220\000\000\006\220\000\000\006\220\000\000\006\220\000"
     "\000\006\220\000\000\006\220\000\000\006\220\000\000\005\120\000" }, 
-  { "trek", 142, 9, 16, 4, 0, 
+  { "trek", NULL, 142, 9, 16, 4, 0, 
     "\001\220\000\124\000\152\100\152\244\152\252\132\232\226\252\244"
     "\152\244\006\244\004\144\106\152\145\246\151\145\226\131\021\226"
     "\104\145\220\031" }, 
-  { "ul_angle", 144, 12, 12, 1, 1, 
+  { "ul_angle", NULL, 144, 12, 12, 1, 1, 
     "\125\125\125\152\252\251\152\252\251\151\125\125\151\000\000\151"
     "\000\000\151\000\000\151\000\000\151\000\000\151\000\000\151\000"
     "\000\125\000\000" }, 
-  { "umbrella", 146, 16, 16, 8, 2, 
+  { "umbrella", NULL, 146, 16, 16, 8, 2, 
     "\001\025\024\124\121\125\125\105\105\225\226\120\025\131\131\225"
     "\145\226\231\140\131\152\245\225\126\006\102\124\000\006\100\000"
     "\000\006\100\000\000\006\100\000\000\006\100\000\000\006\124\000"
     "\000\006\124\000\000\006\144\000\000\006\144\000\000\001\220\000" }, 
-  { "ur_angle", 148, 12, 12, 10, 1, 
+  { "ur_angle", NULL, 148, 12, 12, 10, 1, 
     "\125\125\125\152\252\251\152\252\251\125\125\151\000\000\151\000"
     "\000\151\000\000\151\000\000\151\000\000\151\000\000\151\000\000"
     "\151\000\000\125" }, 
-  { "watch", 150, 16, 16, 15, 9, 
+  { "watch", IDC_WAIT, 150, 16, 16, 15, 9, 
     "\006\252\251\000\006\252\251\000\006\252\251\000\032\252\252\100"
     "\151\126\126\220\245\126\125\245\225\126\125\152\225\132\225\152"
     "\225\132\225\152\225\145\125\152\245\225\125\245\151\125\126\220"
     "\032\252\252\100\006\252\251\000\006\252\251\000\006\252\251\000" }, 
-  { "xterm", 152, 9, 16, 4, 8, 
+  { "xterm", IDC_IBEAM, 152, 9, 16, 4, 8, 
     "\125\025\132\232\225\152\124\026\120\001\220\000\144\000\031\000"
     "\006\100\001\220\000\144\000\031\000\006\100\005\224\025\251\126"
     "\246\245\124\125" }, 
diff -udpr gtk+-2.6.7_/gdk-pixbuf/gdk-pixbuf-private.h gtk+-2.6.7/gdk-pixbuf/gdk-pixbuf-private.h
--- gtk+-2.6.7_/gdk-pixbuf/gdk-pixbuf-private.h	2004-11-16 05:01:57.000000000 +0200
+++ gtk+-2.6.7/gdk-pixbuf/gdk-pixbuf-private.h	2020-12-27 16:22:52.532875169 +0200
@@ -97,9 +97,9 @@ GdkPixbuf *_gdk_pixbuf_generic_image_loa
 GdkPixbufFormat *_gdk_pixbuf_get_format (GdkPixbufModule *image_module);
 
 #ifdef USE_GMODULE
-#define MODULE_ENTRY(type,function) function
+#define MODULE_ENTRY(type,function) __declspec(dllexport) function
 #else
-#define MODULE_ENTRY(type,function) _gdk_pixbuf__ ## type ## _ ## function
+#define MODULE_ENTRY(type,function) __declspec(dllexport) _gdk_pixbuf__ ## type ## _ ## function
 #endif
 
 #endif /* GDK_PIXBUF_ENABLE_BACKEND */
diff -udpr gtk+-2.6.7_/gdk-pixbuf/io-bmp.c gtk+-2.6.7/gdk-pixbuf/io-bmp.c
--- gtk+-2.6.7_/gdk-pixbuf/io-bmp.c	2005-04-10 00:53:39.000000000 +0300
+++ gtk+-2.6.7/gdk-pixbuf/io-bmp.c	2020-12-27 16:22:52.532875169 +0200
@@ -1074,14 +1074,10 @@ gdk_pixbuf__bmp_image_load_increment(gpo
 
 	gint BytesToCopy;
 
-	g_print ("load_inc\n");
 	if (context->read_state == READ_STATE_DONE)
 		return TRUE;
 	else if (context->read_state == READ_STATE_ERROR)
-	{
-		g_print ("ah\n");
 		return FALSE;
-	}
 
 	while (size > 0) {
 		if (context->BufferDone < context->BufferSize) {	/* We still
@@ -1107,37 +1103,25 @@ gdk_pixbuf__bmp_image_load_increment(gpo
 			if (!DecodeHeader (context->buff,
 					   context->buff + 14, context,
 					   error))
-			{
-				g_print ("bla\n");
 				return FALSE;
-			}
 
 			break;
 
 		case READ_STATE_PALETTE:
 			if (!DecodeColormap (context->buff, context, error))
-			{
-				g_print ("bleh\n");
 				return FALSE;
-			}
 			break;
 
 		case READ_STATE_BITMASKS:
 			if (!decode_bitmasks (context->buff, context, error))
-			{
-				g_print ("blurb\n");
 				return FALSE;
-			}
 			break;
 
 		case READ_STATE_DATA:
 			if (context->Compressed == BI_RGB || context->Compressed == BI_BITFIELDS)
 				OneLine (context);
 			else if (!DoCompressed (context, error))
-			{
-				g_print ("blurb\n");
 				return FALSE;
-			}
 
 			break;
 		case READ_STATE_DONE:
diff -udpr gtk+-2.6.7_/gdk-pixbuf/io-pnm.c gtk+-2.6.7/gdk-pixbuf/io-pnm.c
--- gtk+-2.6.7_/gdk-pixbuf/io-pnm.c	2005-02-16 05:39:27.000000000 +0200
+++ gtk+-2.6.7/gdk-pixbuf/io-pnm.c	2020-12-27 16:22:52.532875169 +0200
@@ -245,7 +245,7 @@ pnm_read_next_value (PnmIOBuffer *inbuf,
 	
 	/* get the value */
 	result = strtol (buf, &endptr, 10);
-	if (*endptr != '\0' || result < 0) {
+	if (*endptr != '\0' || result < 0 || result > G_MAXUINT) {
 		g_set_error (error,
 			     GDK_PIXBUF_ERROR,
 			     GDK_PIXBUF_ERROR_CORRUPT_IMAGE,
diff -udpr gtk+-2.6.7_/gdk-pixbuf/io-xpm.c gtk+-2.6.7/gdk-pixbuf/io-xpm.c
--- gtk+-2.6.7_/gdk-pixbuf/io-xpm.c	2005-03-03 16:03:53.000000000 +0200
+++ gtk+-2.6.7/gdk-pixbuf/io-xpm.c	2020-12-27 16:22:52.533875169 +0200
@@ -1167,7 +1167,8 @@ file_buffer (enum buf_op op, gpointer ha
 		/* Fall through to the xpm_read_string. */
 
 	case op_body:
-		xpm_read_string (h->infile, &h->buffer, &h->buffer_size);
+		if(!xpm_read_string (h->infile, &h->buffer, &h->buffer_size))
+			return NULL;
 		return h->buffer;
 
 	default:
@@ -1262,7 +1263,9 @@ pixbuf_create_from_xpm (const gchar * (*
                              _("XPM has invalid number of chars per pixel"));
 		return NULL;
 	}
-	if (n_col <= 0 || n_col >= G_MAXINT / (cpp + 1)) {
+	if (n_col <= 0 || 
+	    n_col >= G_MAXINT / (cpp + 1) || 
+	    n_col >= G_MAXINT / sizeof (XPMColor)) {
                 g_set_error (error,
                              GDK_PIXBUF_ERROR,
                              GDK_PIXBUF_ERROR_CORRUPT_IMAGE,
diff -udpr gtk+-2.6.7_/gtk/gtkcombobox.c gtk+-2.6.7/gtk/gtkcombobox.c
--- gtk+-2.6.7_/gtk/gtkcombobox.c	2005-04-09 01:03:53.000000000 +0300
+++ gtk+-2.6.7/gtk/gtkcombobox.c	2020-12-27 16:22:52.533875169 +0200
@@ -103,6 +103,7 @@ struct _GtkComboBoxPrivate
   guint changed_id;
   guint popup_idle_id;
   guint scroll_timer;
+  guint resize_idle_id;
 
   gint width;
   GSList *cells;
@@ -2852,6 +2853,8 @@ list_popup_resize_idle (gpointer user_da
       gtk_window_move (GTK_WINDOW (combo_box->priv->popup_window), x, y);
     }
 
+  combo_box->priv->resize_idle_id = 0;
+
   GDK_THREADS_LEAVE ();
 
   return FALSE;
@@ -2860,7 +2863,9 @@ list_popup_resize_idle (gpointer user_da
 static void
 gtk_combo_box_list_popup_resize (GtkComboBox *combo_box)
 {
-  g_idle_add (list_popup_resize_idle, combo_box);
+  if (!combo_box->priv->resize_idle_id)
+    combo_box->priv->resize_idle_id = 
+      g_idle_add (list_popup_resize_idle, combo_box);
 }
 
 static void
@@ -3330,6 +3335,12 @@ gtk_combo_box_list_destroy (GtkComboBox
       combo_box->priv->scroll_timer = 0;
     }
 
+  if (combo_box->priv->resize_idle_id)
+    {
+      g_source_remove (combo_box->priv->resize_idle_id);
+      combo_box->priv->resize_idle_id = 0;
+    }
+
   gtk_widget_destroy (combo_box->priv->tree_view);
 
   combo_box->priv->tree_view = NULL;
@@ -3416,7 +3427,8 @@ gtk_combo_box_list_button_released (GtkW
 
   if (ewidget != combo_box->priv->tree_view)
     {
-      if (ewidget == combo_box->priv->button &&
+      if ((ewidget == combo_box->priv->button ||
+	   ewidget == combo_box->priv->box ) &&
           !popup_in_progress &&
           gtk_toggle_button_get_active (GTK_TOGGLE_BUTTON (combo_box->priv->button)))
         {
@@ -3425,7 +3437,8 @@ gtk_combo_box_list_button_released (GtkW
         }
 
       /* released outside treeview */
-      if (ewidget != combo_box->priv->button)
+      if (ewidget != combo_box->priv->button &&
+	  ewidget != combo_box->priv->box)
         {
           gtk_combo_box_popdown (combo_box);
 
@@ -3618,10 +3631,10 @@ gtk_combo_box_list_auto_scroll (GtkCombo
 	  value = adj->value - (tree_view->allocation.y - y + 1);
 	  gtk_adjustment_set_value (adj, CLAMP (value, adj->lower, adj->upper - adj->page_size));
 	}
-      else if (y >= tree_view->allocation.y + tree_view->allocation.height &&
+      else if (y >= tree_view->allocation.height &&
 	       adj->upper - adj->page_size > adj->value)
 	{
-	  value = adj->value + (y - tree_view->allocation.y - tree_view->allocation.height + 1);
+	  value = adj->value + (y - tree_view->allocation.height + 1);
 	  gtk_adjustment_set_value (adj, CLAMP (value, 0.0, adj->upper - adj->page_size));
 	}
     }
diff -udpr gtk+-2.6.7_/gtk/gtkdnd.c gtk+-2.6.7/gtk/gtkdnd.c
--- gtk+-2.6.7_/gtk/gtkdnd.c	2005-03-20 08:25:27.000000000 +0200
+++ gtk+-2.6.7/gtk/gtkdnd.c	2020-12-27 16:22:52.534875169 +0200
@@ -1949,11 +1949,7 @@ gtk_drag_begin_internal (GtkWidget
       return NULL;
     }
 
-  if (gdk_keyboard_grab (ipc_widget->window, FALSE, time) != 0)
-    {
-      gtk_drag_release_ipc_widget (ipc_widget);
-      return NULL;
-    }
+  gdk_keyboard_grab (ipc_widget->window, FALSE, time);
 
   /* We use a GTK grab here to override any grabs that the widget
    * we are dragging from might have held
@@ -2145,6 +2141,9 @@ gtk_drag_source_set (GtkWidget
       g_signal_connect (widget, "button_press_event",
 			G_CALLBACK (gtk_drag_source_event_cb),
 			site);
+      g_signal_connect (widget, "button_release_event",
+			G_CALLBACK (gtk_drag_source_event_cb),
+			site);
       g_signal_connect (widget, "motion_notify_event",
 			G_CALLBACK (gtk_drag_source_event_cb),
 			site);
@@ -2183,9 +2182,6 @@ gtk_drag_source_unset (GtkWidget
       g_signal_handlers_disconnect_by_func (widget,
 					    gtk_drag_source_event_cb,
 					    site);
-      g_signal_handlers_disconnect_by_func (widget,
-					    gtk_drag_source_event_cb,
-					    site);
       g_object_set_data (G_OBJECT (widget), "gtk-site-data", NULL);
     }
 }
diff -udpr gtk+-2.6.7_/gtk/gtkentry.c gtk+-2.6.7/gtk/gtkentry.c
--- gtk+-2.6.7_/gtk/gtkentry.c	2005-04-07 07:45:30.000000000 +0300
+++ gtk+-2.6.7/gtk/gtkentry.c	2020-12-27 16:22:52.534875169 +0200
@@ -75,6 +75,8 @@ struct _GtkEntryPrivate
 {
   gfloat xalign;
   gint insert_pos;
+  guint real_changed : 1;
+  guint change_count : 8;
 };
 
 enum {
@@ -349,6 +351,9 @@ static void         disconnect_completio
 static void         connect_completion_signals         (GtkEntry           *entry,
 							GtkEntryCompletion *completion);
 
+static void         begin_change                       (GtkEntry *entry);
+static void         end_change                         (GtkEntry *entry);
+static void         emit_changed                       (GtkEntry *entry);
 
 static GtkWidgetClass *parent_class = NULL;
 
@@ -1024,6 +1029,46 @@ gtk_entry_init (GtkEntry *entry)
 		    G_CALLBACK (gtk_entry_delete_surrounding_cb), entry);
 }
 
+static void
+begin_change (GtkEntry *entry)
+{
+  GtkEntryPrivate *priv = GTK_ENTRY_GET_PRIVATE (entry);
+
+  priv->change_count++;
+}
+
+static void
+end_change (GtkEntry *entry)
+{
+  GtkEditable *editable = GTK_EDITABLE (entry);
+  GtkEntryPrivate *priv = GTK_ENTRY_GET_PRIVATE (entry);
+ 
+  g_return_if_fail (priv->change_count > 0);
+
+  priv->change_count--;
+
+  if (priv->change_count == 0)
+    {
+       if (priv->real_changed) 
+         {
+           g_signal_emit_by_name (editable, "changed");
+           priv->real_changed = FALSE;
+         }
+    } 
+}
+
+static void
+emit_changed (GtkEntry *entry)
+{
+  GtkEditable *editable = GTK_EDITABLE (entry);
+  GtkEntryPrivate *priv = GTK_ENTRY_GET_PRIVATE (entry);
+
+  if (priv->change_count == 0)
+    g_signal_emit_by_name (editable, "changed");
+  else 
+    priv->real_changed = TRUE;
+}
+
 /*
  * Overwrite a memory that might contain sensitive information.
  */
@@ -2243,7 +2288,7 @@ gtk_entry_real_insert_text (GtkEditable
 
   gtk_entry_recompute (entry);
 
-  g_signal_emit_by_name (editable, "changed");
+  emit_changed (entry);
   g_object_notify (G_OBJECT (editable), "text");
 }
 
@@ -2293,7 +2338,7 @@ gtk_entry_real_delete_text (GtkEditable
       
       gtk_entry_recompute (entry);
 
-      g_signal_emit_by_name (editable, "changed");
+      emit_changed (entry);
       g_object_notify (G_OBJECT (editable), "text");
     }
 }
@@ -3553,7 +3598,7 @@ gtk_entry_move_forward_word (GtkEntry *e
       
       /* Find the next word end */
       new_pos++;
-      while (new_pos < n_attrs && !log_attrs[new_pos].is_word_end)
+      while (new_pos < n_attrs - 1 && !log_attrs[new_pos].is_word_end)
 	new_pos++;
 
       g_free (log_attrs);
@@ -3688,12 +3733,16 @@ paste_received (GtkClipboard *clipboard,
 	    _gtk_entry_completion_popdown (completion);
 	}
 
+      begin_change (entry);
+      g_object_freeze_notify (G_OBJECT (entry));
       if (gtk_editable_get_selection_bounds (editable, &start, &end))
         gtk_editable_delete_text (editable, start, end);
 
       pos = entry->current_pos;
       gtk_editable_insert_text (editable, text, -1, &pos);
       gtk_editable_set_position (editable, pos);
+      g_object_thaw_notify (G_OBJECT (entry));
+      end_change (entry);
 
       if (completion)
 	g_signal_handler_unblock (entry, completion->priv->changed_id);
@@ -3838,10 +3887,14 @@ gtk_entry_set_text (GtkEntry    *entry,
   if (completion)
     g_signal_handler_block (entry, completion->priv->changed_id);
 
+  begin_change (entry);
+  g_object_freeze_notify (G_OBJECT (entry));
   gtk_editable_delete_text (GTK_EDITABLE (entry), 0, -1);
 
   tmp_pos = 0;
   gtk_editable_insert_text (GTK_EDITABLE (entry), text, strlen (text), &tmp_pos);
+  g_object_thaw_notify (G_OBJECT (entry));
+  end_change (entry);
 
   if (completion)
     g_signal_handler_unblock (entry, completion->priv->changed_id);
@@ -4724,8 +4777,12 @@ gtk_entry_drag_data_received (GtkWidget
       else
 	{
 	  /* Replacing selection */
+          begin_change (entry);
+          g_object_freeze_notify (G_OBJECT (entry));
 	  gtk_editable_delete_text (editable, sel1, sel2);
 	  gtk_editable_insert_text (editable, str, -1, &sel1);
+          g_object_thaw_notify (G_OBJECT (entry));
+          end_change (entry);
 	}
       
       g_free (str);
@@ -4790,6 +4847,7 @@ cursor_blinks (GtkEntry *entry)
   gboolean blink;
 
   if (GTK_WIDGET_HAS_FOCUS (entry) &&
+      entry->editable &&
       entry->selection_bound == entry->current_pos)
     {
       g_object_get (settings, "gtk-cursor-blink", &blink, NULL);
diff -udpr gtk+-2.6.7_/gtk/gtkfilesel.c gtk+-2.6.7/gtk/gtkfilesel.c
--- gtk+-2.6.7_/gtk/gtkfilesel.c	2005-04-07 07:45:30.000000000 +0300
+++ gtk+-2.6.7/gtk/gtkfilesel.c	2020-12-27 16:22:52.535875169 +0200
@@ -2070,7 +2070,8 @@ win32_gtk_add_drives_to_dir_list (GtkLis
   while (*textPtr != '\0')
     {
       /* Ignore floppies (?) */
-      if (GetDriveType (textPtr) != DRIVE_REMOVABLE)
+// We need to see flash drives too - WJ
+//      if (GetDriveType (textPtr) != DRIVE_REMOVABLE)
 	{
 	  /* Build the actual displayable string */
 	  g_snprintf (formatBuffer, sizeof (formatBuffer), "%c:\\", toupper (textPtr[0]));
diff -udpr gtk+-2.6.7_/gtk/gtkmain.c gtk+-2.6.7/gtk/gtkmain.c
--- gtk+-2.6.7_/gtk/gtkmain.c	2005-03-20 08:25:28.000000000 +0200
+++ gtk+-2.6.7/gtk/gtkmain.c	2020-12-27 16:22:52.535875169 +0200
@@ -290,13 +290,46 @@ _gtk_get_libdir (void)
   return gtk_libdir;
 }
 
+/* Lifted from HEAD GLib */
+static gchar *
+g_win32_locale_filename_from_utf8 (const gchar *utf8filename)
+{
+  gchar *retval = g_locale_from_utf8 (utf8filename, -1, NULL, NULL, NULL);
+
+  if (retval == NULL && G_WIN32_HAVE_WIDECHAR_API ())
+    {
+      /* Conversion failed, so convert to wide chars, check if there
+       * is a 8.3 version, and use that.
+       */
+      wchar_t *wname = g_utf8_to_utf16 (utf8filename, -1, NULL, NULL, NULL);
+      if (wname != NULL)
+	{
+	  wchar_t wshortname[MAX_PATH + 1];
+	  if (GetShortPathNameW (wname, wshortname, G_N_ELEMENTS (wshortname)))
+	    {
+	      gchar *tem = g_utf16_to_utf8 (wshortname, -1, NULL, NULL, NULL);
+	      retval = g_locale_from_utf8 (tem, -1, NULL, NULL, NULL);
+	      g_free (tem);
+	    }
+	  g_free (wname);
+	}
+    }
+  return retval;
+}
+
 const gchar *
 _gtk_get_localedir (void)
 {
   static char *gtk_localedir = NULL;
   if (gtk_localedir == NULL)
-    gtk_localedir = g_win32_get_package_installation_subdirectory
-      (GETTEXT_PACKAGE, dll_name, "lib\\locale");
+    {
+      gtk_localedir = g_win32_get_package_installation_subdirectory (GETTEXT_PACKAGE, dll_name, "lib\\locale");
+      if (gtk_localedir != NULL)
+	gtk_localedir = g_win32_locale_filename_from_utf8 (gtk_localedir);
+
+      if (gtk_localedir == NULL)
+	gtk_localedir = "\\";	/* Punt */
+    }
 
   return gtk_localedir;
 }
diff -udpr gtk+-2.6.7_/gtk/gtknotebook.c gtk+-2.6.7/gtk/gtknotebook.c
--- gtk+-2.6.7_/gtk/gtknotebook.c	2005-04-07 07:45:30.000000000 +0300
+++ gtk+-2.6.7/gtk/gtknotebook.c	2020-12-27 16:22:52.536875169 +0200
@@ -4370,10 +4370,8 @@ gtk_notebook_insert_page_menu (GtkNotebo
   if (!notebook->first_tab)
     notebook->first_tab = notebook->children;
 
-  if (!notebook->cur_page)
-    gtk_widget_set_child_visible (child, TRUE);
-  else
-    gtk_widget_set_child_visible (child, FALSE);
+  /* child visible will be turned on by switch_page below */
+  gtk_widget_set_child_visible (child, FALSE);
   
   if (tab_label)
     {
diff -udpr gtk+-2.6.7_/gtk/gtkradiobutton.c gtk+-2.6.7/gtk/gtkradiobutton.c
--- gtk+-2.6.7_/gtk/gtkradiobutton.c	2005-03-20 08:25:28.000000000 +0200
+++ gtk+-2.6.7/gtk/gtkradiobutton.c	2020-12-27 16:22:52.536875169 +0200
@@ -510,7 +510,7 @@ gtk_radio_button_focus (GtkWidget
 	    {
 	      GtkWidget *child = tmp_list->data;
 	      
-	      if (GTK_WIDGET_VISIBLE (child) && GTK_WIDGET_IS_SENSITIVE (child))
+	      if (GTK_WIDGET_REALIZED (child) && GTK_WIDGET_IS_SENSITIVE (child))
 		{
 		  new_focus = child;
 		  break;
@@ -528,7 +528,7 @@ gtk_radio_button_focus (GtkWidget
 	    {
 	      GtkWidget *child = tmp_list->data;
 	      
-	      if (GTK_WIDGET_VISIBLE (child) && GTK_WIDGET_IS_SENSITIVE (child))
+	      if (GTK_WIDGET_REALIZED (child) && GTK_WIDGET_IS_SENSITIVE (child))
 		{
 		  new_focus = child;
 		  break;
diff -udpr gtk+-2.6.7_/gtk/gtkscale.c gtk+-2.6.7/gtk/gtkscale.c
--- gtk+-2.6.7_/gtk/gtkscale.c	2005-04-07 07:45:30.000000000 +0300
+++ gtk+-2.6.7/gtk/gtkscale.c	2020-12-27 16:22:52.536875169 +0200
@@ -265,10 +265,10 @@ gtk_scale_class_init (GtkScaleClass *cla
   add_slider_binding (binding_set, GDK_KP_Down, GDK_CONTROL_MASK,
                       GTK_SCROLL_PAGE_DOWN);
    
-  add_slider_binding (binding_set, GDK_Page_Up, 0,
+  add_slider_binding (binding_set, GDK_Page_Up, GDK_CONTROL_MASK,
                       GTK_SCROLL_PAGE_LEFT);
 
-  add_slider_binding (binding_set, GDK_KP_Page_Up, 0,
+  add_slider_binding (binding_set, GDK_KP_Page_Up, GDK_CONTROL_MASK,
                       GTK_SCROLL_PAGE_LEFT);  
 
   add_slider_binding (binding_set, GDK_Page_Up, 0,
@@ -277,10 +277,10 @@ gtk_scale_class_init (GtkScaleClass *cla
   add_slider_binding (binding_set, GDK_KP_Page_Up, 0,
                       GTK_SCROLL_PAGE_UP);
   
-  add_slider_binding (binding_set, GDK_Page_Down, 0,
+  add_slider_binding (binding_set, GDK_Page_Down, GDK_CONTROL_MASK,
                       GTK_SCROLL_PAGE_RIGHT);
 
-  add_slider_binding (binding_set, GDK_KP_Page_Down, 0,
+  add_slider_binding (binding_set, GDK_KP_Page_Down, GDK_CONTROL_MASK,
                       GTK_SCROLL_PAGE_RIGHT);
 
   add_slider_binding (binding_set, GDK_Page_Down, 0,
diff -udpr gtk+-2.6.7_/gtk/gtkwindow.c gtk+-2.6.7/gtk/gtkwindow.c
--- gtk+-2.6.7_/gtk/gtkwindow.c	2005-04-07 08:09:33.000000000 +0300
+++ gtk+-2.6.7/gtk/gtkwindow.c	2020-12-27 16:22:52.537875169 +0200
@@ -4695,16 +4695,20 @@ gtk_window_real_set_focus (GtkWindow *wi
 {
   GtkWidget *old_focus = window->focus_widget;
   gboolean had_default = FALSE;
+  gboolean focus_had_default = FALSE;
+  gboolean old_focus_had_default = FALSE;
 
   if (old_focus)
     {
       g_object_ref (old_focus);
       g_object_freeze_notify (G_OBJECT (old_focus));
+      old_focus_had_default = GTK_WIDGET_HAS_DEFAULT (old_focus);
     }
   if (focus)
     {
       g_object_ref (focus);
       g_object_freeze_notify (G_OBJECT (focus));
+      focus_had_default = GTK_WIDGET_HAS_DEFAULT (focus);
     }
   
   if (window->default_widget)
@@ -4716,10 +4720,11 @@ gtk_window_real_set_focus (GtkWindow *wi
 	  (window->focus_widget != window->default_widget))
         {
 	  GTK_WIDGET_UNSET_FLAGS (window->focus_widget, GTK_HAS_DEFAULT);
-
+	  gtk_widget_queue_draw (window->focus_widget);
+	  
 	  if (window->default_widget)
 	    GTK_WIDGET_SET_FLAGS (window->default_widget, GTK_HAS_DEFAULT);
-        }
+	}
 
       window->focus_widget = NULL;
 
@@ -4761,14 +4766,20 @@ gtk_window_real_set_focus (GtkWindow *wi
   if (window->default_widget &&
       (had_default != GTK_WIDGET_HAS_DEFAULT (window->default_widget)))
     gtk_widget_queue_draw (window->default_widget);
-
+  
   if (old_focus)
     {
+      if (old_focus_had_default != GTK_WIDGET_HAS_DEFAULT (old_focus))
+	gtk_widget_queue_draw (old_focus);
+	
       g_object_thaw_notify (G_OBJECT (old_focus));
       g_object_unref (old_focus);
     }
   if (focus)
     {
+      if (focus_had_default != GTK_WIDGET_HAS_DEFAULT (focus))
+	gtk_widget_queue_draw (focus);
+
       g_object_thaw_notify (G_OBJECT (focus));
       g_object_unref (focus);
     }
diff -udpr gtk+-2.6.7_/modules/engines/ms-windows/msw_style.c gtk+-2.6.7/modules/engines/ms-windows/msw_style.c
--- gtk+-2.6.7_/modules/engines/ms-windows/msw_style.c	2005-03-21 00:40:58.000000000 +0200
+++ gtk+-2.6.7/modules/engines/ms-windows/msw_style.c	2020-12-27 16:22:52.537875169 +0200
@@ -1448,10 +1448,13 @@ draw_box (GtkStyle      *style,
     }
   else if (detail && strcmp (detail, "menuitem") == 0) {
     shadow_type = GTK_SHADOW_NONE;
+#if 0
+    /* don't work with Vista */
       if (xp_theme_draw (window, XP_THEME_ELEMENT_MENU_ITEM, style, x, y, width, height, state_type, area))
         {
   		return;
         }
+#endif
   }
   else if (detail && !strcmp (detail, "trough"))
     {
